<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Configuration Page &mdash; PathLay 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=2709fde1"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Maps Databases" href="usage_access_mapsdb.html" />
    <link rel="prev" title="Downloading an Experiment Package" href="../home/usage_home_download.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            PathLay
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../indices/indices_index.html">Welcome to PathLay’s documentation!</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../features/index.html">Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../requirements/index.html">Requirements</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../requirements/requirements_docker.html">For Building the Docker Image (Recommended)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../requirements/requirements_apache.html">For the Apache Server Installation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../installation/index.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../installation/linux/index.html">Installation on Linux</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../installation/linux/installation_docker.html">Installation using Docker (Recommended)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../installation/linux/installation_apache.html">Apache Server Installation</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../howto/index.html">How to run</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../howto/howto_docker_server.html">Run PathLay as network server</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../usage_index.html">Usage</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../account/usage_account_index.html">Account Managing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../account/usage_account_new.html">Register a New Account</a></li>
<li class="toctree-l3"><a class="reference internal" href="../account/usage_account_login.html">Login to an Existing Account</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../exp_package/usage_exp_package_index.html">Experiment Package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../exp_package/usage_exp_package_conf.html">Configuration File</a></li>
<li class="toctree-l3"><a class="reference internal" href="../exp_package/usage_exp_package_specs_data.html">Datasets Files</a></li>
<li class="toctree-l3"><a class="reference internal" href="../exp_package/usage_exp_package_specs_ids.html">Supported IDs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../home/usage_home_index.html">Home Manager</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../home/usage_home_create.html">Creating an Experiment Package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../home/usage_home_load.html">Loading an Experiment Package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../home/usage_home_ont.html">Gene Ontologies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../home/usage_home_download.html">Downloading an Experiment Package</a></li>
</ul>
</li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Configuration Page</a><ul>
<li class="toctree-l3"><a class="reference internal" href="usage_access_mapsdb.html">Maps Databases</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage_access_stat.html">Maps Restriction Procedure</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage_access_data.html">Data Selection</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage_access_data_types.html">Configuring a Dataset</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage_access_filtering.html">Filtering Options</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage_access_id_only.html">ID Preservation</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage_access_nodeg.html">Non Differentially Expressed IDs</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage_access_tfs.html">Transcription Factors</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage_access_tfs_id_only.html">Transcription Factors ID Preservation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../results/usage_results_index.html">Pathway Explorer</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../results/usage_results_mapsel.html">Pathway Navigation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../results/usage_results_legend.html">Indicators</a></li>
<li class="toctree-l3"><a class="reference internal" href="../results/usage_results_menu.html">Main Menu</a></li>
<li class="toctree-l3"><a class="reference internal" href="../results/usage_results_highlight.html">Highlighting Feature</a></li>
<li class="toctree-l3"><a class="reference internal" href="../results/usage_results_clipboard.html">The Clipboard</a></li>
<li class="toctree-l3"><a class="reference internal" href="../results/selector/usage_results_selector.html">The Selectors Section</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../results/selector/usage_results_select_by_ids.html">Select by IDs</a></li>
<li class="toctree-l4"><a class="reference internal" href="../results/selector/usage_results_select_by_agreement.html">Select by Agreement</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../results/settings/usage_results_settings.html">Settings</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../results/settings/usage_results_settings_transparency_and_size.html">Transparency and Size</a></li>
<li class="toctree-l4"><a class="reference internal" href="../results/settings/usage_results_settings_screenshot.html">Screenshot</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../license/index.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../support/index.html">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contribute/index.html">Contribute</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">PathLay</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../usage_index.html">Usage</a></li>
      <li class="breadcrumb-item active">Configuration Page</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/usage/access/usage_access_index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="configuration-page">
<span id="usage-access"></span><h1>Configuration Page<a class="headerlink" href="#configuration-page" title="Link to this heading"></a></h1>
<div class="toctree-wrapper compound">
</div>
<p>The Configuration Page is accessible from the Login Page by clicking “Access PathLay” once you input your credentials. From this page you will be able to run PathLay’s analysis on an Experiment Package.
First thing to do is select an Experiment Package and load it using the dropdown menu and the “Load” button.</p>
<section id="maps-databases">
<span id="usage-access-mapdb"></span><h2>Maps Databases<a class="headerlink" href="#maps-databases" title="Link to this heading"></a></h2>
<p>The “Maps Database” section allows you to choose which database will be used to load pathways from, it currently provides two possibilities, both for Homo sapiens pathways: KEGG and WikiPathways.</p>
</section>
<section id="maps-restriction-procedure">
<span id="usage-access-stat"></span><h2>Maps Restriction Procedure<a class="headerlink" href="#maps-restriction-procedure" title="Link to this heading"></a></h2>
<p>It is possible to perform a selection of maps to display in the PathLay - MAPS page by selecting FET from the Map Restriction Procedure selector.
Maps are selected with a statistical approach based on the Fisher’s Exact Test and since PathLay integrates different datasets, it is also possible to select which dataset will contribute on the map selection.
To allow this, the “FET Enabled” option should be checked for each dataset that should contribute.
By default the test is performed on each dataset enabled and it will produce a list of maps considered of importance, all the lists produced are then joined into one.
There are two tweaks available that can change the result of this step: the Intersect option and the Pooling option.</p>
<section id="pooling">
<h3>Pooling<a class="headerlink" href="#pooling" title="Link to this heading"></a></h3>
<p>Eabling the Pooling option will affect how the Transcriptomic, Proteomic, miRNomic, Methylomic and Cromatin Status datasets are treated when enabled for the test.
When enabled, the IDs of the above datasets will be merged and the test is performed on that single list.</p>
</section>
<section id="intersect">
<h3>Intersect<a class="headerlink" href="#intersect" title="Link to this heading"></a></h3>
<p>Enabling the Intersect option will affect which maps will be selected after the test has been performed.
By default map lists are joined into one, meaning that even if a map has been found relevant in relation to just one of the datasets, it will be kept and indicators will be plotted on it.
With the Intersect feature enabled, only maps that are considered relevant for all the datasets enabled are kept (i.e. the ones that appear in every map list coming from the tests performed).</p>
</section>
</section>
<section id="configuring-a-dataset">
<span id="usage-data-types"></span><h2>Configuring a Dataset<a class="headerlink" href="#configuring-a-dataset" title="Link to this heading"></a></h2>
<p>Under the “Select Maps using” section is located a selector that allows to switch between the configurations available for each dataset.
Some of these configurations are common to every data type, while others, related to the Transcription Factors (TFs) and the Non Differentially Expressed components (NoDE), are peculiar to a few of them.</p>
</section>
<section id="filtering-options">
<span id="usage-data-filters"></span><h2>Filtering Options<a class="headerlink" href="#filtering-options" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>There are three filters available for each data type, which can be enabled by checking the respective box and requires a threshold value to be written in the input field next to it. These three filters are summarized below:</dt><dd><ul class="simple">
<li><p>Effect Size “&lt;” : filters out all the components with an Effect Size value greater than the threshold selected</p></li>
<li><p>Effect Size “&gt;” : filters out all the components with an Effect Size value smaller than the threshold selected</p></li>
<li><p>p-value “&lt;” : filters out all the components with a p-value value greater than the threshold</p></li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Whenever a threshold field displays a red background, it means that the aforementioned threshold is not valid and contains a typing error. In this situation if the related filter checkbox remains enabled, the submit button will disappear until this error is either fixed by changing the threshold value or by disabling the filter.</p>
</div>
</section>
<section id="usage-data-id-only">
<span id="id-preservation"></span><h2>ID Preservation<a class="headerlink" href="#usage-data-id-only" title="Link to this heading"></a></h2>
<p>ID Preservation can be enabled for every data type by checking the “Preserve non DE IDs” checkbox.
This feature will not let PathLay to discard IDs found in the dataset without an Effect Size value.
These “ID Only” components will be represented with a different palette of colors since they do not provide any information regarding their differential expression, but they will be still integrated with the other datasets.</p>
</section>
<section id="non-differentially-expressed-ids">
<span id="usage-access-nodeg"></span><h2>Non Differentially Expressed IDs<a class="headerlink" href="#non-differentially-expressed-ids" title="Link to this heading"></a></h2>
<p>Whenever a miRNa, Methylation or Chromatin status dataset is enabled, miRNas, methylations and chromatin statuses will be represented alongside the related genes.
PathLays’s approach is to integrate the information held within different datasets whenever is possible, hence when a Transcriptomic or Proteomic dataset is provided and enabled alongside the afromentioned ones, the standard procedure adopted is to represent an integration of these informations using genes and proteins provided as a scaffold on which miRNAs, methylations and chromatin statuses impact.
Anyway It is possible to represent miRNas, methylations or chromatin statuses despite the absence of their related genes in the Transcriptomic dataset, by checking the “No DE Loading” feature in their respective configurations.
This feature allows PathLay to display miRnas, methylations and chromatin statuses on grey indicators that represent a non-differentially expressed gene (i.e. a gene not provided in other datasets).
This feature specifically works with the IDs of miRNAs, methylations and chromatin statuses linked to an Effect Size value, while the IDs without an Effect Size value will be only linked and represented alongside those genes or proteins that are effectively provided by their respectiv datasets, unless the “No DE Loading From Preserved IDs” feature is enabled.
By doing this, ID only components coming from these thre datasets will also be able to call out Non DE genes and proteins.
Graphically speaking this is translated into grey indicators surrounded by smaller circles and squares coloured in orange.</p>
</section>
<section id="transcription-factors">
<span id="usage-access-tfs"></span><h2>Transcription Factors<a class="headerlink" href="#transcription-factors" title="Link to this heading"></a></h2>
<p>PathLay also supports GTRD as a database and can display transcription factors (TFs) that interacts with genes coming from a Transcriptomic or Proteomic dataset and as well recognize if those gene/protein IDs actually refers directly to a transcription factor.
Transcription factors will be loaded and linked to the gene IDs in the Transcriptomic dataset if the “Enable TFs” feature is enabled. The same goes for the proteomic dataset which has its own “Enable TFs” feature.
If a transcription factor is found in the Transcriptomic dataset it will be linked to any related gene also present in the dataset.
Whenever Transcriptomic and Proteomic datasets are enabled together and the “Enable TFs” feature is enabled for at least one of them, once a TF is recognized, its targets will be looked for both in the Transcriptomic and Proteomic dataset to guarantee an high level degree of integration.</p>
</section>
<section id="transcription-factors-id-preservation">
<span id="usage-access-id-only-tfs"></span><h2>Transcription Factors ID Preservation<a class="headerlink" href="#transcription-factors-id-preservation" title="Link to this heading"></a></h2>
<p>Normally, if a Transcription Factor is recognized in an ID Only component found in the Transcriptomic or Proteomic dataset (when of course the “Preserve non DE IDs” feature is enabled for them), it is just loaded as a DE component and not represented as a TF.
To overcome this, one should enable the “Preserve IDs for TFs” feature in the Gene or Protein configuration depending on the scope of the analysis.
Transcription Factors recognized in this manner will be represented as small oragne squares on the right side of the main indicators.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../home/usage_home_download.html" class="btn btn-neutral float-left" title="Downloading an Experiment Package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="usage_access_mapsdb.html" class="btn btn-neutral float-right" title="Maps Databases" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Matteo Ramazzotti Lorenzo Casbarra.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-XXXXXXXXXX"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-XXXXXXXXXX', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>