<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage &mdash; PathLay 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=2709fde1"></script>
        <script src="../_static/doctools.js?v=9a2dae69"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Account Managing" href="account/usage_account_index.html" />
    <link rel="prev" title="Run PathLay as network server" href="../howto/howto_docker_server.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            PathLay
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../indices/indices_index.html">Welcome to PathLay’s documentation!</a></li>
<li class="toctree-l1"><a class="reference internal" href="../features/index.html">Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="../requirements/index.html">Requirements</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../requirements/requirements_docker.html">For Building the Docker Image (Recommended)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../requirements/requirements_apache.html">For the Apache Server Installation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation/index.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation/linux/index.html">Installation on Linux</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../installation/linux/installation_docker.html">Installation using Docker (Recommended)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../installation/linux/installation_apache.html">Apache Server Installation</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../howto/index.html">How to run</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../howto/howto_docker_server.html">Run PathLay as network server</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="account/usage_account_index.html">Account Managing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="account/usage_account_new.html">Register a New Account</a></li>
<li class="toctree-l3"><a class="reference internal" href="account/usage_account_login.html">Login to an Existing Account</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="exp_package/usage_exp_package_index.html">Experiment Package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="exp_package/usage_exp_package_conf.html">Configuration File</a></li>
<li class="toctree-l3"><a class="reference internal" href="exp_package/usage_exp_package_specs_data.html">Datasets Files</a></li>
<li class="toctree-l3"><a class="reference internal" href="exp_package/usage_exp_package_specs_ids.html">Supported IDs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="home/usage_home_index.html">Home Manager</a><ul>
<li class="toctree-l3"><a class="reference internal" href="home/usage_home_create.html">Creating an Experiment Package</a></li>
<li class="toctree-l3"><a class="reference internal" href="home/usage_home_load.html">Loading an Experiment Package</a></li>
<li class="toctree-l3"><a class="reference internal" href="home/usage_home_ont.html">Gene Ontologies</a></li>
<li class="toctree-l3"><a class="reference internal" href="home/usage_home_download.html">Downloading an Experiment Package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="access/usage_access_index.html">Configuration Page</a><ul>
<li class="toctree-l3"><a class="reference internal" href="access/usage_access_mapsdb.html">Maps Databases</a></li>
<li class="toctree-l3"><a class="reference internal" href="access/usage_access_stat.html">Maps Restriction Procedure</a></li>
<li class="toctree-l3"><a class="reference internal" href="access/usage_access_data.html">Data Selection</a></li>
<li class="toctree-l3"><a class="reference internal" href="access/usage_access_data_types.html">Configuring a Dataset</a></li>
<li class="toctree-l3"><a class="reference internal" href="access/usage_access_filtering.html">Filtering Options</a></li>
<li class="toctree-l3"><a class="reference internal" href="access/usage_access_id_only.html">ID Preservation</a></li>
<li class="toctree-l3"><a class="reference internal" href="access/usage_access_nodeg.html">Non Differentially Expressed IDs</a></li>
<li class="toctree-l3"><a class="reference internal" href="access/usage_access_tfs.html">Transcription Factors</a></li>
<li class="toctree-l3"><a class="reference internal" href="access/usage_access_tfs_id_only.html">Transcription Factors ID Preservation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="results/usage_results_index.html">Pathway Explorer</a><ul>
<li class="toctree-l3"><a class="reference internal" href="results/usage_results_mapsel.html">Pathway Navigation</a></li>
<li class="toctree-l3"><a class="reference internal" href="results/usage_results_legend.html">Indicators</a></li>
<li class="toctree-l3"><a class="reference internal" href="results/usage_results_menu.html">Main Menu</a></li>
<li class="toctree-l3"><a class="reference internal" href="results/usage_results_highlight.html">Highlighting Feature</a></li>
<li class="toctree-l3"><a class="reference internal" href="results/usage_results_clipboard.html">The Clipboard</a></li>
<li class="toctree-l3"><a class="reference internal" href="results/selector/usage_results_selector.html">The Selectors Section</a><ul>
<li class="toctree-l4"><a class="reference internal" href="results/selector/usage_results_select_by_ids.html">Select by IDs</a></li>
<li class="toctree-l4"><a class="reference internal" href="results/selector/usage_results_select_by_agreement.html">Select by Agreement</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="results/settings/usage_results_settings.html">Settings</a><ul>
<li class="toctree-l4"><a class="reference internal" href="results/settings/usage_results_settings_transparency_and_size.html">Transparency and Size</a></li>
<li class="toctree-l4"><a class="reference internal" href="results/settings/usage_results_settings_screenshot.html">Screenshot</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../license/index.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../support/index.html">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute/index.html">Contribute</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">PathLay</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Usage</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/usage/usage_index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="usage">
<span id="id1"></span><h1>Usage<a class="headerlink" href="#usage" title="Link to this heading"></a></h1>
<div class="toctree-wrapper compound">
</div>
<section id="account-managing">
<span id="usage-account"></span><h2>Account Managing<a class="headerlink" href="#account-managing" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
</div>
</section>
<section id="experiment-package">
<span id="usage-specs"></span><h2>Experiment Package<a class="headerlink" href="#experiment-package" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
</div>
<p>In the following section, the structure of a zipped Experiment Package will be explained. Please note that each and avery file listed below can be created and modified from your Home Page direclty, with no need to manually adjust them from outside PathLay.
Each Experiment Package is defined from up to 7 different text files with different extensions:</p>
<table class="docutils align-default" id="id4">
<caption><span class="caption-text">Experiment Package Content</span><a class="headerlink" href="#id4" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>File Extension</p></th>
<th class="head"><p>Content</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>.conf</p></td>
<td><p>Holds the main configuration of the Experiment Package</p></td>
</tr>
<tr class="row-odd"><td><p>.mrna</p></td>
<td><p>Holds the Transcriptomic dataset</p></td>
</tr>
<tr class="row-even"><td><p>.prot</p></td>
<td><p>Holds the Proteomic dataset</p></td>
</tr>
<tr class="row-odd"><td><p>.mirna</p></td>
<td><p>Holds the miRNomic dataset</p></td>
</tr>
<tr class="row-even"><td><p>.meth</p></td>
<td><p>Holds the Methylomic dataset</p></td>
</tr>
<tr class="row-odd"><td><p>.chroma</p></td>
<td><p>Holds the Chromatin Status dataset</p></td>
</tr>
<tr class="row-even"><td><p>.ont</p></td>
<td><p>Holds the Gene Ontologies entries of interest of the Experiment Package</p></td>
</tr>
<tr class="row-odd"><td><p>.meta</p></td>
<td><p>Holds the Metabolomic dataset</p></td>
</tr>
</tbody>
</table>
<p>Every experiment created in your Home Page has a name like “exp1,2,3…” automatically assigned. If you want to manually configure an experiment from scratch you have to name your files with this convention in mind (i.e. exp1.conf, exp1.mrna, exp1.meta etc…).</p>
<section id="configuration-file">
<span id="usage-specs-conf"></span><h3>Configuration File<a class="headerlink" href="#configuration-file" title="Link to this heading"></a></h3>
<p>The configuration file for the Experiment Package is structured as a series of tag and value pairs, separated by a “=” character, with each line of the file holding one pair.</p>
<table class="docutils align-default" id="id5">
<caption><span class="caption-text">Configuration File Tags</span><a class="headerlink" href="#id5" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 15.8%" />
<col style="width: 52.6%" />
<col style="width: 31.6%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Tag</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Content</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>expname</p></td>
<td><p>Holds the name of the Experiment Package</p></td>
<td><p>String of text</p></td>
</tr>
<tr class="row-odd"><td><p>comments</p></td>
<td><p>Holds brief comments for the Experiment Package</p></td>
<td><p>String of text</p></td>
</tr>
<tr class="row-even"><td><p>organism</p></td>
<td><p>Holds the organism ID related to the experiment</p></td>
<td><p>hsa; mmu</p></td>
</tr>
<tr class="row-odd"><td><p>geneIdType</p></td>
<td><p>Holds the ID type for the Transcriptomic dataset</p></td>
<td><p>entrez; ensembl; symbol;</p></td>
</tr>
<tr class="row-even"><td><p>gene_id_column</p></td>
<td><p>Holds the number representing the column with the IDs in the Transcriptomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>gene_dev_column</p></td>
<td><p>Holds the number representing the column with the differential expression values in the Transcriptomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>gene_pvalue_column</p></td>
<td><p>Holds the number representing the column with the p-values in the Transcriptomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>urnaIdType</p></td>
<td><p>Holds the ID type for the miRNomic dataset</p></td>
<td><p>mirbase</p></td>
</tr>
<tr class="row-even"><td><p>urna_id_column</p></td>
<td><p>Holds the number representing the column with the IDs in the miRNomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>urna_dev_column</p></td>
<td><p>Holds the number representing the column with the differential expression values in the miRNomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>urna_pvalue_column</p></td>
<td><p>Holds the number representing the column with the p-values in the miRNomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>metaIdType</p></td>
<td><p>Holds the ID type for the Metabolomic dataset</p></td>
<td><p>keggcompound; name;</p></td>
</tr>
<tr class="row-even"><td><p>meta_id_column</p></td>
<td><p>Holds the number representing the column with the IDs in the Metabolomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>meta_dev_column</p></td>
<td><p>Holds the number representing the column with the expression values in the Metabolomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>meta_pvalue_column</p></td>
<td><p>Holds the number representing the column with the p-values in the Metabolomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>methIdType</p></td>
<td><p>Holds the ID type for the Methylomic dataset</p></td>
<td><p>entrez; ensembl; symbol;</p></td>
</tr>
<tr class="row-even"><td><p>meth_id_column</p></td>
<td><p>Holds the number representing the column with the IDs in the Methylomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>meth_dev_column</p></td>
<td><p>Holds the number representing the column with the differential expression values in the Methylomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>meth_pvalue_column</p></td>
<td><p>Holds the number representing the column with the p-values in the Methylomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>protIdType</p></td>
<td><p>Holds the ID type for the Proteomic dataset</p></td>
<td><p>entry; entrez; symbol;</p></td>
</tr>
<tr class="row-even"><td><p>prot_id_column</p></td>
<td><p>Holds the number representing the column with the IDs in the Proteomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>prot_dev_column</p></td>
<td><p>Holds the number representing the column with the differential expression values in the Proteomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>prot_pvalue_column</p></td>
<td><p>Holds the number representing the column with the p-values in the Proteomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>chromaIdType</p></td>
<td><p>Holds the ID type for the Chromatin Status dataset</p></td>
<td><p>entrez; ensembl; symbol;</p></td>
</tr>
<tr class="row-even"><td><p>chroma_id_column</p></td>
<td><p>Holds the number representing the column with the IDs in the Chromatin Status dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>chroma_dev_column</p></td>
<td><p>Holds the number representing the column with the differential expression values in the Chromatin Status dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>chroma_pvalue_column</p></td>
<td><p>Holds the number representing the column with the p-values in the Chromatin Status dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
</tbody>
</table>
<p>An example of a fully configured .conf file is displayed below:</p>
<div class="literal-block-wrapper docutils container" id="id6">
<div class="code-block-caption"><span class="caption-text">Structure of a .conf file of an Experiment Package with Transcriptomic, Proteomic, miRNomic, Methylomic and Metabolomic datasets associated</span><a class="headerlink" href="#id6" title="Link to this code"></a></div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">expname</span><span class="o">=</span><span class="n">LTED</span> <span class="n">vs</span> <span class="n">MCF7</span><span class="o">+</span>
<span class="n">comments</span><span class="o">=</span><span class="n">ANOVA</span> <span class="n">adj</span><span class="o">.</span><span class="n">p</span> <span class="o">+</span> <span class="n">Tukey</span>
<span class="n">organism</span><span class="o">=</span><span class="n">hsa</span>
<span class="n">geneIdType</span><span class="o">=</span><span class="n">entrez</span>
<span class="n">gene_id_column</span><span class="o">=</span><span class="mi">8</span>
<span class="n">gene_dev_column</span><span class="o">=</span><span class="mi">6</span>
<span class="n">gene_pvalue_column</span><span class="o">=</span><span class="mi">5</span>
<span class="n">protIdType</span><span class="o">=</span><span class="n">entry</span>
<span class="n">prot_id_column</span><span class="o">=</span><span class="mi">1</span>
<span class="n">prot_dev_column</span><span class="o">=</span><span class="mi">4</span>
<span class="n">prot_pvalue_column</span><span class="o">=</span><span class="mi">3</span>
<span class="n">urnaIdType</span><span class="o">=</span><span class="n">mirbase</span>
<span class="n">urna_id_column</span><span class="o">=</span><span class="mi">1</span>
<span class="n">urna_dev_column</span><span class="o">=</span><span class="mi">6</span>
<span class="n">urna_pvalue_column</span><span class="o">=</span><span class="mi">5</span>
<span class="n">methIdType</span><span class="o">=</span><span class="n">entrez</span>
<span class="n">meth_id_column</span><span class="o">=</span><span class="mi">8</span>
<span class="n">meth_dev_column</span><span class="o">=</span><span class="mi">6</span>
<span class="n">meth_pvalue_column</span><span class="o">=</span><span class="mi">5</span>
<span class="n">metaIdType</span><span class="o">=</span><span class="n">keggcompound</span>
<span class="n">meta_id_column</span><span class="o">=</span><span class="mi">8</span>
<span class="n">meta_dev_column</span><span class="o">=</span><span class="mi">6</span>
<span class="n">meta_pvalue_column</span><span class="o">=</span><span class="mi">5</span>
<span class="n">chromaIdType</span><span class="o">=</span><span class="n">entrez</span>
<span class="n">chroma_id_column</span><span class="o">=</span><span class="mi">8</span>
<span class="n">chroma_dev_column</span><span class="o">=</span><span class="mi">6</span>
<span class="n">chroma_pvalue_column</span><span class="o">=</span><span class="mi">5</span>
</pre></div>
</div>
</div>
<p>The experiment Package is named “LTED vs MCF7+” and it’s related to an experiment involving Homo sapiens cells, it has all the -omics datasets currently supported by PathLay associated to it.
We can read its configuration as it follows:</p>
<ul class="simple">
<li><p>The Transcriptomic datataset has its Entrez IDs stored in the 8th column, Effect Size values stored in the 6th column and p-values stored in the 5th column.</p></li>
<li><p>The miRNomic datataset has its mirbase IDs stored in the 1st column, Effect Size values stored in the 6th column and p-values stored in the 5th column.</p></li>
<li><p>The Metabolomic datataset has its Kegg Compound IDs stored in the 8th column, Effect Size values stored in the 6th column and p-values stored in the 5th column.</p></li>
<li><p>The Methylomic datataset has its Entrez IDs stored in the 8th column, Effect Size values stored in the 6th column and p-values stored in the 5th column.</p></li>
<li><p>The Chromatin Status dataset has its Entrez IDs stored in the 8th column, Effect Size values stored in the 6th column and p-values stored in the 5th column.</p></li>
<li><p>The Proteomic datataset has its Uniprot Entry IDs stored in the 1st column, Effect Size values stored in the 4th column and p-values stored in the 3rd column.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Experiment Packages can be created and configured in your Home Page with a more intuitive approach. Once the “Save” button is clicked the .conf file will be automatically generated and saved in your home folder.</p>
</div>
</section>
<section id="datasets-files">
<span id="usage-specs-data"></span><h3>Datasets Files<a class="headerlink" href="#datasets-files" title="Link to this heading"></a></h3>
<p>Each dataset file (i.e. a file with .mrna, .prot, .mirna, .meta, .meth and .chroma extension) is a tab separated file that can have any number of columns but at least one, two or three, depending on the Analysis configuration you will choose for PathLay (see more in Configuration Page section), the only mandatory field is the one containing the IDs.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Datasets can be copied and pasted in their related text areas in your Home Page. Once the “Save” button is clicked the dataset file will be created in your home folder and named after the experiment.</p>
</div>
</section>
<section id="supported-ids">
<span id="usage-specs-ids"></span><h3>Supported IDs<a class="headerlink" href="#supported-ids" title="Link to this heading"></a></h3>
<p>PathLay supports a variety of ID types as input which are converted to the ID default type during analysis.
All the ID types are summarized at <a class="reference internal" href="#usage-specs-table-id-compatibility"><span class="std std-ref">Table</span></a> and some examples are provided at <a class="reference internal" href="#usage-specs-table-id-examples"><span class="std std-ref">Table</span></a>.</p>
<span id="usage-specs-table-id-compatibility"></span><table class="docutils align-default" id="id7">
<caption><span class="caption-text">Compatible IDs to use in your Datasets</span><a class="headerlink" href="#id7" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 33.3%" />
<col style="width: 33.3%" />
<col style="width: 33.3%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Dataset Type</p></th>
<th class="head"><p>Supported Input IDs</p></th>
<th class="head"><p>ID Type for Analysis</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Gene</p></td>
<td><p>Entrez Gene ID;  Ensembl; Gene Symbol;</p></td>
<td><p>Entrez Gene ID</p></td>
</tr>
<tr class="row-odd"><td><p>Protein</p></td>
<td><p>Entrez Gene ID; UniProtKB Entry; Protein Symbol;</p></td>
<td><p>Entrez Gene ID</p></td>
</tr>
<tr class="row-even"><td><p>miRNA</p></td>
<td><p>miRBase ID</p></td>
<td><p>miRBase ID</p></td>
</tr>
<tr class="row-odd"><td><p>Metabolite</p></td>
<td><p>KEGG Compounds; Compound Name;</p></td>
<td><p>KEGG Compounds</p></td>
</tr>
<tr class="row-even"><td><p>Methylation</p></td>
<td><p>Entrez Gene ID;  Ensembl; Gene Symbol;</p></td>
<td><p>Entrez Gene ID</p></td>
</tr>
<tr class="row-odd"><td><p>Chromatin Status</p></td>
<td><p>Entrez Gene ID;  Ensembl; Gene Symbol;</p></td>
<td><p>Entrez Gene ID</p></td>
</tr>
</tbody>
</table>
<span id="usage-specs-table-id-examples"></span><table class="docutils align-default" id="id8">
<caption><span class="caption-text">Examples of ID Types</span><a class="headerlink" href="#id8" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 50.0%" />
<col style="width: 50.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID Type</p></th>
<th class="head"><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Entrez Gene ID</p></td>
<td><p>2308</p></td>
</tr>
<tr class="row-odd"><td><p>Gene Symbol</p></td>
<td><p>FOXO1</p></td>
</tr>
<tr class="row-even"><td><p>Ensembl</p></td>
<td><p>ENSG00000150907</p></td>
</tr>
<tr class="row-odd"><td><p>UniProtKB Entry</p></td>
<td><p>Q12778</p></td>
</tr>
<tr class="row-even"><td><p>Protein Symbol</p></td>
<td><p>FOXO1</p></td>
</tr>
<tr class="row-odd"><td><p>miRBase ID</p></td>
<td><p>hsa-miR-183-5p</p></td>
</tr>
<tr class="row-even"><td><p>KEGG Compounds</p></td>
<td><p>C03758</p></td>
</tr>
<tr class="row-odd"><td><p>Compound Name</p></td>
<td><p>Dopamine</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="home-manager">
<span id="usage-home"></span><h2>Home Manager<a class="headerlink" href="#home-manager" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
</div>
<p>Once an account is registered the Home Page becomes available, while for an already registered account it is accessible by clicking the “Manage Home” button in the login page. In the Home Page you can create, load and modify an Experiment Package with a more intuitive approach.</p>
<section id="creating-an-experiment-package">
<span id="usage-home-create"></span><h3>Creating an Experiment Package<a class="headerlink" href="#creating-an-experiment-package" title="Link to this heading"></a></h3>
<p>To create an Experiment Package from scratch the first thing to do is to click the “Add New” button and then assign it a name in the “Experiment Title” text box, you can also comment it in the “Comments” text box located below.
Next, you should assign it an organism between hsa and mmu from the Organism selector.
You can now proceed to load your datasets on PathLay.
Data types supported may come from Transcriptomic (Genes), miRNomic (miRNAs), Metabolomic (Metabolites), Methylomic (Methylations), Proteomic (Proteins) and Chromatin Status experiments.
Each of the aforementioned data types have a coloured and manually resizable text box in which is possible to copy and paste a tab separated dataset (see “Dataset Files”):</p>
<ul class="simple">
<li><p>Genes datasets can be loaded in the light blue text box</p></li>
<li><p>miRNAs datasets can be loaded in the red text box</p></li>
<li><p>Metabolites datasets can be loaded in the yellow text box</p></li>
<li><p>Methylations datasets can be loaded in the green text box</p></li>
<li><p>Proteins datasets can be loaded in the purple text box</p></li>
<li><p>Chromatin Status datasets can be loaded in the purple text box</p></li>
</ul>
<p>Only one text box is displayed at a time, to change this simply click on the dropdown menu “Data Type” and select one of the six data types.
When a dataset is displayed, its configuration becomes available for you to see and set up.
This setup requires you the following:</p>
<ul class="simple">
<li><p>Select the ID Type you are submitting from the “ID Type” selector, each data type has its own suppported IDs</p></li>
<li><p>Mark the column number of the dataset that stores the IDs using the “ID Column” text box</p></li>
<li><p>Mark the column number of the dataset that stores the differential expression values using the text box next to the aforementioned dropdown menu</p></li>
<li><p>Mark the column number of the dataset that stores p-values and write a threshold for them</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Depending on the Analysis you will setup later for PathLay, not all of the above configurations must be performed, onle the ID Type and the ID Column are mandatory.</p>
</div>
<p>Once the configuration is done you can click the “Save” button and the Experiment Package will be created.
The “Save” button will of course also save any modification you will submit to PathLay when you will reload this Experiment Package in the future.</p>
</section>
<section id="loading-an-experiment-package">
<span id="usage-home-load"></span><h3>Loading an Experiment Package<a class="headerlink" href="#loading-an-experiment-package" title="Link to this heading"></a></h3>
<p>To load an existing Experiment Package from your home folder, select its name from the dropdown menu in the top left corner and hit the “Load” button. Datasets and configurations will be displayed in their relative fields.
If you have a zipped Experiment Package and you want to load it on PathLay, simply click the “Browse” button in the top right corner and select the zipped archive, then click the “Upload” button and load it.</p>
</section>
<section id="gene-ontologies">
<span id="usage-home-ont"></span><h3>Gene Ontologies<a class="headerlink" href="#gene-ontologies" title="Link to this heading"></a></h3>
<p>Your Home Page also provides a list of cellular compartments that are nothing less than Gene Ontologies.
Each one of the entries represents a gene list of Entrez Gene IDs and you can select the ones you are more interested in by clicking the “+” button located on the right side of each box.
A search box is also provided to filter the ones you are most interested in.
Gene Ontologies of interest will be added to the other list alongside the preselected ones: Nucleus, Cytoplasm, Mitochondrion, Lysosome, Endoplasmic Reticulum, Golgi Apparatus, Cytoskeleton, Membrane.
You can of course remove them by clicking the “-” button in the selected boxes.
Cellular compartments selected will be found in a dropdown menu after PathLay has completed the analysis and it’s ready to show the results on functional pathways.
Once a cellular compartment is chosen, only genes and proteins localized in that compartment will remain displayed on the pathways (see more “Pathway Explorer”).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The selection of Gene Ontologies is as well part of your configuration for the Experiment Package, so any change you provide will be saved with the “Save” button.</p>
</div>
</section>
<section id="downloading-an-experiment-package">
<span id="usage-home-download"></span><h3>Downloading an Experiment Package<a class="headerlink" href="#downloading-an-experiment-package" title="Link to this heading"></a></h3>
<p>The “Download” button allows you to package the currently loaded datasets and configurations in a zip archive and export it from PathLay for other uses, for instance, this archive can be uploaded as a new Experiment Package into another PathLay account.
You can also download the entirety of your home folder (i.e. all the datasets and configurations of all the experiments) with the “Download Home” button.</p>
</section>
</section>
<section id="configuration-page">
<span id="usage-access"></span><h2>Configuration Page<a class="headerlink" href="#configuration-page" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
</div>
<p>The Configuration Page is accessible from the Login Page by clicking “Access PathLay” once you input your credentials. From this page you will be able to run PathLay’s analysis on an Experiment Package.
First thing to do is select an Experiment Package and load it using the dropdown menu and the “Load” button.</p>
<section id="maps-databases">
<span id="usage-access-mapdb"></span><h3>Maps Databases<a class="headerlink" href="#maps-databases" title="Link to this heading"></a></h3>
<p>The “Maps Database” section allows you to choose which database will be used to load pathways from, it currently provides two possibilities, both for Homo sapiens pathways: KEGG and WikiPathways.</p>
</section>
<section id="maps-restriction-procedure">
<span id="usage-access-stat"></span><h3>Maps Restriction Procedure<a class="headerlink" href="#maps-restriction-procedure" title="Link to this heading"></a></h3>
<p>It is possible to perform a selection of maps to display in the PathLay - MAPS page by selecting FET from the Map Restriction Procedure selector.
Maps are selected with a statistical approach based on the Fisher’s Exact Test and since PathLay integrates different datasets, it is also possible to select which dataset will contribute on the map selection.
To allow this, the “FET Enabled” option should be checked for each dataset that should contribute.
By default the test is performed on each dataset enabled and it will produce a list of maps considered of importance, all the lists produced are then joined into one.
There are two tweaks available that can change the result of this step: the Intersect option and the Pooling option.</p>
<section id="pooling">
<h4>Pooling<a class="headerlink" href="#pooling" title="Link to this heading"></a></h4>
<p>Eabling the Pooling option will affect how the Transcriptomic, Proteomic, miRNomic, Methylomic and Cromatin Status datasets are treated when enabled for the test.
When enabled, the IDs of the above datasets will be merged and the test is performed on that single list.</p>
</section>
<section id="intersect">
<h4>Intersect<a class="headerlink" href="#intersect" title="Link to this heading"></a></h4>
<p>Enabling the Intersect option will affect which maps will be selected after the test has been performed.
By default map lists are joined into one, meaning that even if a map has been found relevant in relation to just one of the datasets, it will be kept and indicators will be plotted on it.
With the Intersect feature enabled, only maps that are considered relevant for all the datasets enabled are kept (i.e. the ones that appear in every map list coming from the tests performed).</p>
</section>
</section>
<section id="data-selection">
<span id="usage-access-data"></span><h3>Data Selection<a class="headerlink" href="#data-selection" title="Link to this heading"></a></h3>
<p>The “Select Maps using” section allows to select the datasets you want to use by simply check the boxes you are interested in between “Genes”, “Proteins”, “miRNAs”, “Methylations”, “Chromatin Status” and “Metabolites”.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Whenever a checkbox is disabled, it means that either the dataset was not found available or that the setup in the Home page was not performed accordingly (i.e. the content of the columns was not pointed out correctly).</p>
</div>
</section>
<section id="configuring-a-dataset">
<span id="usage-data-types"></span><h3>Configuring a Dataset<a class="headerlink" href="#configuring-a-dataset" title="Link to this heading"></a></h3>
<p>Under the “Select Maps using” section is located a selector that allows to switch between the configurations available for each dataset.
Some of these configurations are common to every data type, while others, related to the Transcription Factors (TFs) and the Non Differentially Expressed components (NoDE), are peculiar to a few of them.</p>
</section>
<section id="filtering-options">
<span id="usage-data-filters"></span><h3>Filtering Options<a class="headerlink" href="#filtering-options" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>There are three filters available for each data type, which can be enabled by checking the respective box and requires a threshold value to be written in the input field next to it. These three filters are summarized below:</dt><dd><ul class="simple">
<li><p>Effect Size “&lt;” : filters out all the components with an Effect Size value greater than the threshold selected</p></li>
<li><p>Effect Size “&gt;” : filters out all the components with an Effect Size value smaller than the threshold selected</p></li>
<li><p>p-value “&lt;” : filters out all the components with a p-value value greater than the threshold</p></li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Whenever a threshold field displays a red background, it means that the aforementioned threshold is not valid and contains a typing error. In this situation if the related filter checkbox remains enabled, the submit button will disappear until this error is either fixed by changing the threshold value or by disabling the filter.</p>
</div>
</section>
<section id="usage-data-id-only">
<span id="id-preservation"></span><h3>ID Preservation<a class="headerlink" href="#usage-data-id-only" title="Link to this heading"></a></h3>
<p>ID Preservation can be enabled for every data type by checking the “Preserve non DE IDs” checkbox.
This feature will not let PathLay to discard IDs found in the dataset without an Effect Size value.
These “ID Only” components will be represented with a different palette of colors since they do not provide any information regarding their differential expression, but they will be still integrated with the other datasets.</p>
</section>
<section id="non-differentially-expressed-ids">
<span id="usage-access-nodeg"></span><h3>Non Differentially Expressed IDs<a class="headerlink" href="#non-differentially-expressed-ids" title="Link to this heading"></a></h3>
<p>Whenever a miRNa, Methylation or Chromatin status dataset is enabled, miRNas, methylations and chromatin statuses will be represented alongside the related genes.
PathLays’s approach is to integrate the information held within different datasets whenever is possible, hence when a Transcriptomic or Proteomic dataset is provided and enabled alongside the afromentioned ones, the standard procedure adopted is to represent an integration of these informations using genes and proteins provided as a scaffold on which miRNAs, methylations and chromatin statuses impact.
Anyway It is possible to represent miRNas, methylations or chromatin statuses despite the absence of their related genes in the Transcriptomic dataset, by checking the “No DE Loading” feature in their respective configurations.
This feature allows PathLay to display miRnas, methylations and chromatin statuses on grey indicators that represent a non-differentially expressed gene (i.e. a gene not provided in other datasets).
This feature specifically works with the IDs of miRNAs, methylations and chromatin statuses linked to an Effect Size value, while the IDs without an Effect Size value will be only linked and represented alongside those genes or proteins that are effectively provided by their respectiv datasets, unless the “No DE Loading From Preserved IDs” feature is enabled.
By doing this, ID only components coming from these thre datasets will also be able to call out Non DE genes and proteins.
Graphically speaking this is translated into grey indicators surrounded by smaller circles and squares coloured in orange.</p>
</section>
<section id="transcription-factors">
<span id="usage-access-tfs"></span><h3>Transcription Factors<a class="headerlink" href="#transcription-factors" title="Link to this heading"></a></h3>
<p>PathLay also supports GTRD as a database and can display transcription factors (TFs) that interacts with genes coming from a Transcriptomic or Proteomic dataset and as well recognize if those gene/protein IDs actually refers directly to a transcription factor.
Transcription factors will be loaded and linked to the gene IDs in the Transcriptomic dataset if the “Enable TFs” feature is enabled. The same goes for the proteomic dataset which has its own “Enable TFs” feature.
If a transcription factor is found in the Transcriptomic dataset it will be linked to any related gene also present in the dataset.
Whenever Transcriptomic and Proteomic datasets are enabled together and the “Enable TFs” feature is enabled for at least one of them, once a TF is recognized, its targets will be looked for both in the Transcriptomic and Proteomic dataset to guarantee an high level degree of integration.</p>
</section>
<section id="transcription-factors-id-preservation">
<span id="usage-access-id-only-tfs"></span><h3>Transcription Factors ID Preservation<a class="headerlink" href="#transcription-factors-id-preservation" title="Link to this heading"></a></h3>
<p>Normally, if a Transcription Factor is recognized in an ID Only component found in the Transcriptomic or Proteomic dataset (when of course the “Preserve non DE IDs” feature is enabled for them), it is just loaded as a DE component and not represented as a TF.
To overcome this, one should enable the “Preserve IDs for TFs” feature in the Gene or Protein configuration depending on the scope of the analysis.
Transcription Factors recognized in this manner will be represented as small oragne squares on the right side of the main indicators.</p>
</section>
<section id="id2">
<span id="id3"></span><h3>Non Differentially Expressed IDs<a class="headerlink" href="#id2" title="Link to this heading"></a></h3>
<p>Whenever a miRNa, Methylation or Chromatin status dataset is enabled, miRNas, methylations and chromatin statuses will be represented alongside the related genes.
PathLays’s approach is to integrate the information held within different datasets whenever is possible, hence when a Transcriptomic or Proteomic dataset is provided and enabled alongside the afromentioned ones, the standard procedure adopted is to represent an integration of these informations using genes and proteins provided as a scaffold on which miRNAs, methylations and chromatin statuses impact.
Anyway It is possible to represent miRNas, methylations or chromatin statuses despite the absence of their related genes in the Transcriptomic dataset, by checking the “No DE Loading” feature in their respective configurations.
This feature allows PathLay to display miRnas, methylations and chromatin statuses on grey indicators that represent a non-differentially expressed gene (i.e. a gene not provided in other datasets).
This feature specifically works with the IDs of miRNAs, methylations and chromatin statuses linked to an Effect Size value, while the IDs without an Effect Size value will be only linked and represented alongside those genes or proteins that are effectively provided by their respectiv datasets, unless the “No DE Loading From Preserved IDs” feature is enabled.
By doing this, ID only components coming from these thre datasets will also be able to call out Non DE genes and proteins.
Graphically speaking this is translated into grey indicators surrounded by smaller circles and squares coloured in orange.</p>
</section>
</section>
<section id="pathway-explorer">
<span id="usage-results"></span><h2>Pathway Explorer<a class="headerlink" href="#pathway-explorer" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
</div>
<p>The “PathLay - MAPS” page will be displayed after the configuration and the submitting of an analysis on an Experiment Package in your home. IDs coming from your datasets will be portrayed as different indicators on pathways depending on the Analysis Mode chosen (see more “Selecting an Analysis Mode”). This page allows you to:</p>
<blockquote>
<div><ul class="simple">
<li><p>Display your datasets on pathways</p></li>
<li><p>Filter pathways by selecting a specific gene/protein/miRNA/TF ID or by multiple IDs</p></li>
<li><p>Select a cellular compartment and display only the IDs localized in it</p></li>
<li><p>Select indicators and display their infos on the clipboard</p></li>
<li><p>Save indicators of interest and download them as a tab-separated text file</p></li>
<li><p>Filter pathways by selecting an “Agreement” between Gene/Proteins and Methylation, miRNAs or TFs</p></li>
</ul>
</div></blockquote>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The “PathLay - MAPS” page is directly accessible after you hit the “Submit” button allowing you to navigate through pathways right away but depending on how much data PathLay needs to process, some indicators may not be displayed correctly from the go and require more time.</p>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../howto/howto_docker_server.html" class="btn btn-neutral float-left" title="Run PathLay as network server" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="account/usage_account_index.html" class="btn btn-neutral float-right" title="Account Managing" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Matteo Ramazzotti Lorenzo Casbarra.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-XXXXXXXXXX"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-XXXXXXXXXX', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>