<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage &mdash; PathLay 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=2709fde1"></script>
        <script src="../_static/doctools.js?v=9a2dae69"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Account Managing" href="account/usage_account_index.html" />
    <link rel="prev" title="Run PathLay as local server with Docker (localhost)" href="../howto/howto_localserver.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            PathLay
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../indices/indices_index.html">Welcome to PathLay’s documentation!</a></li>
<li class="toctree-l1"><a class="reference internal" href="../features/index.html">Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="../requirements/index.html">Requirements</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../requirements/requirements_docker.html">For Building the Docker Image (Recommended)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../requirements/requirements_apache.html">For the Apache Server Installation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation/installation_index.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation/linux/installation_linux_index.html">Installation on Linux</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../installation/linux/installation_docker.html">Installation using Docker (Recommended)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../installation/linux/installation_deb.html">Installation with .deb package</a></li>
<li class="toctree-l3"><a class="reference internal" href="../installation/linux/installation_apache.html">Apache Server Installation</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../howto/index.html">How to run</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../howto/howto_docker_server.html">Run PathLay as network server with Docker</a></li>
<li class="toctree-l2"><a class="reference internal" href="../howto/howto_localserver.html">Run PathLay as local server with Docker (localhost)</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="account/usage_account_index.html">Account Managing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="account/usage_account_new.html">Register a New Account</a></li>
<li class="toctree-l3"><a class="reference internal" href="account/usage_account_login.html">Login to an Existing Account</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="exp_package/usage_exp_package_index.html">Experiment Package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="exp_package/usage_exp_package_conf.html">Configuration File</a></li>
<li class="toctree-l3"><a class="reference internal" href="exp_package/usage_exp_package_specs_data.html">Datasets Files</a></li>
<li class="toctree-l3"><a class="reference internal" href="exp_package/usage_exp_package_specs_ids.html">Supported IDs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="dbconf/dbconf_index.html">Database Configuration</a><ul>
<li class="toctree-l3"><a class="reference internal" href="dbconf/dbconf_org_selection.html">Organism Selection</a></li>
<li class="toctree-l3"><a class="reference internal" href="dbconf/dbconf_interactions.html">Interaction Files</a></li>
<li class="toctree-l3"><a class="reference internal" href="dbconf/dbconf_pathways.html">Pathway Files</a></li>
<li class="toctree-l3"><a class="reference internal" href="dbconf/dbconf_backup_repo.html">Alternative installation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="home/usage_home_index.html">Home Manager</a><ul>
<li class="toctree-l3"><a class="reference internal" href="home/create/usage_home_create.html">Creating an Experiment Package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="home/create/usage_home_create_add.html">Adding a new Experiment Package</a></li>
<li class="toctree-l4"><a class="reference internal" href="home/create/usage_home_create_upload.html">Uploading a new Experiment Package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="home/edit/usage_home_edit_index.html">Editing an Experiment Package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="home/edit/usage_home_edit_generic.html">Edit Experiment’s General Information</a></li>
<li class="toctree-l4"><a class="reference internal" href="home/edit/usage_home_edit_dataset.html">Configure a Dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="home/edit/usage_home_edit_ont.html">Gene Ontologies</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="home/usage_home_download.html">Downloading an Experiment Package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="access/usage_access_index.html">Configuration Page</a><ul>
<li class="toctree-l3"><a class="reference internal" href="access/usage_access_mapsdb.html">Maps Databases</a></li>
<li class="toctree-l3"><a class="reference internal" href="access/usage_access_data.html">Data Selection</a></li>
<li class="toctree-l3"><a class="reference internal" href="access/filters/usage_access_filters_index.html">Configuring Filters for a Dataset</a><ul>
<li class="toctree-l4"><a class="reference internal" href="access/filters/usage_access_filtering.html">Filtering Options</a></li>
<li class="toctree-l4"><a class="reference internal" href="access/filters/usage_access_id_only.html">ID Preservation</a></li>
<li class="toctree-l4"><a class="reference internal" href="access/filters/tfs/usage_access_tfs.html">Transcription Factors</a><ul>
<li class="toctree-l5"><a class="reference internal" href="access/filters/tfs/usage_access_tfs_id_only.html">Transcription Factors ID Preservation</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="access/filters/node/usage_access_nodeg.html">Non Differentially Expressed IDs</a><ul>
<li class="toctree-l5"><a class="reference internal" href="access/filters/node/usage_access_nodeg_tfs.html">Loading Non Differentially Expressed IDs from TFs</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="access/stat/usage_access_stat.html">Maps Restriction Procedure</a><ul>
<li class="toctree-l4"><a class="reference internal" href="access/stat/usage_access_stat_pool.html">Pooling</a></li>
<li class="toctree-l4"><a class="reference internal" href="access/stat/usage_access_stat_intersect.html">Intersect</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="results/usage_results_index.html">Pathway Explorer</a><ul>
<li class="toctree-l3"><a class="reference internal" href="results/usage_results_mapsel.html">Pathway Navigation</a></li>
<li class="toctree-l3"><a class="reference internal" href="results/usage_results_legend.html">Indicators</a></li>
<li class="toctree-l3"><a class="reference internal" href="results/usage_results_menu.html">Main Menu</a></li>
<li class="toctree-l3"><a class="reference internal" href="results/usage_results_highlight.html">Highlighting Feature</a></li>
<li class="toctree-l3"><a class="reference internal" href="results/usage_results_clipboard.html">The Clipboard</a></li>
<li class="toctree-l3"><a class="reference internal" href="results/selector/usage_results_selector.html">The Selectors Section</a><ul>
<li class="toctree-l4"><a class="reference internal" href="results/selector/usage_results_select_by_ids.html">Select by IDs</a></li>
<li class="toctree-l4"><a class="reference internal" href="results/selector/usage_results_select_by_agreement.html">Select by Agreement</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="results/settings/usage_results_settings.html">Settings</a><ul>
<li class="toctree-l4"><a class="reference internal" href="results/settings/usage_results_settings_transparency_and_size.html">Transparency and Size</a></li>
<li class="toctree-l4"><a class="reference internal" href="results/settings/usage_results_settings_screenshot.html">Screenshot</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../license/index.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../support/index.html">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute/index.html">Contribute</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">PathLay</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Usage</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/usage/usage_index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="usage">
<span id="id1"></span><h1>Usage<a class="headerlink" href="#usage" title="Link to this heading"></a></h1>
<div class="toctree-wrapper compound">
</div>
<section id="account-managing">
<span id="usage-account"></span><h2>Account Managing<a class="headerlink" href="#account-managing" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
</div>
</section>
<section id="experiment-package">
<span id="usage-specs"></span><h2>Experiment Package<a class="headerlink" href="#experiment-package" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
</div>
<p>In the following section, the structure of a zipped Experiment Package will be explained. Please note that each and avery file listed below can be created and modified from your Home Page direclty, with no need to manually adjust them from outside PathLay.
Each Experiment Package is defined from up to 7 different text files with different extensions:</p>
<table class="docutils align-default" id="id2">
<caption><span class="caption-text">Experiment Package Content</span><a class="headerlink" href="#id2" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>File Extension</p></th>
<th class="head"><p>Content</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>.conf</p></td>
<td><p>Holds the main configuration of the Experiment Package</p></td>
</tr>
<tr class="row-odd"><td><p>.mrna</p></td>
<td><p>Holds the Transcriptomic dataset</p></td>
</tr>
<tr class="row-even"><td><p>.prot</p></td>
<td><p>Holds the Proteomic dataset</p></td>
</tr>
<tr class="row-odd"><td><p>.mirna</p></td>
<td><p>Holds the miRNomic dataset</p></td>
</tr>
<tr class="row-even"><td><p>.meth</p></td>
<td><p>Holds the Methylomic dataset</p></td>
</tr>
<tr class="row-odd"><td><p>.chroma</p></td>
<td><p>Holds the Chromatin Status dataset</p></td>
</tr>
<tr class="row-even"><td><p>.ont</p></td>
<td><p>Holds the Gene Ontologies entries of interest of the Experiment Package</p></td>
</tr>
<tr class="row-odd"><td><p>.meta</p></td>
<td><p>Holds the Metabolomic dataset</p></td>
</tr>
</tbody>
</table>
<p>Every experiment created in your Home Page has a name like “exp1,2,3…” automatically assigned. If you want to manually configure an experiment from scratch you have to name your files with this convention in mind (i.e. exp1.conf, exp1.mrna, exp1.meta etc…).</p>
</section>
<section id="database-configuration">
<span id="usage-dbconf"></span><h2>Database Configuration<a class="headerlink" href="#database-configuration" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
</div>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/welcome_settings.png"><img alt="access" src="../_images/welcome_settings.png" style="width: 100%;" /></a>
</figure>
<p>Database configuration is accessible only from the host machine where PathLay is installed, this is due to avoid that many users in parallel can run it.
It is a crucial step to be performed to make PathLay usable and it articulates as follows:</p>
<ol class="arabic simple">
<li><p>Select the organism to Configure</p></li>
<li><p>Check for files regarding components interactions</p></li>
<li><p>Check for pathway files</p></li>
</ol>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>There could be problems during the installation.
If that happens please refer to the <a class="reference internal" href="#dbconf-backup-repo"><span class="std std-ref">Alternative installation</span></a> section.</p>
</div>
<section id="organism-selection">
<span id="dbconf-org-selection"></span><h3>Organism Selection<a class="headerlink" href="#organism-selection" title="Link to this heading"></a></h3>
<p>To select the organism database to configure use the selector on the top right.
At this stage PathLay supports only Homo Sapiens and Mus musculus.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/dbconf_select_org.png"><img alt="access" src="../_images/dbconf_select_org.png" style="width: 100%;" /></a>
</figure>
</section>
<section id="interaction-files">
<span id="dbconf-org-interactions"></span><h3>Interaction Files<a class="headerlink" href="#interaction-files" title="Link to this heading"></a></h3>
<p>Press the “Check File Existance” button to check if the database is ready.
On your first installation of PathLay of course this will result in all files missing, as reported by the red boxes.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/dbconf_check_file.png"><img alt="check_files_button" src="../_images/dbconf_check_file.png" style="width: 100%;" /></a>
</figure>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/dbconf_files_fail.png"><img alt="check_fail" src="../_images/dbconf_files_fail.png" style="width: 100%;" /></a>
</figure>
<p>To begin the installation process click the “Download” button and simply wait for the downloads to finish.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/dbconf_files_success.png"><img alt="check_success" src="../_images/dbconf_files_success.png" style="width: 100%;" /></a>
</figure>
<p>Once all the downloads are done, the boxes will turn green and you can proceed to the next step.</p>
</section>
<section id="pathway-files">
<span id="dbconf-org-pathways"></span><h3>Pathway Files<a class="headerlink" href="#pathway-files" title="Link to this heading"></a></h3>
<p>Press the “Check Maps Existance” button to check the pathway database status.
Pathway databases supported are KEGG and WikiPathways.
On your first installation of PathLay of course this will result in all files missing, as reported by the red boxes.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/dbconf_check_maps.png"><img alt="conf_check_maps" src="../_images/dbconf_check_maps.png" style="width: 100%;" /></a>
</figure>
<p>To begin the installation process click the “Download” button and simply wait for the downloads to finish.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/dbconf_maps.png"><img alt="conf_maps" src="../_images/dbconf_maps.png" style="width: 60%;" /></a>
</figure>
<p>You can monitor which maps are being downloaded in the corresponding box.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Downloading may take some time, and it is recommended to download one database at a time for a smoother process.</p>
</div>
</section>
<section id="alternative-installation">
<span id="dbconf-backup-repo"></span><h3>Alternative installation<a class="headerlink" href="#alternative-installation" title="Link to this heading"></a></h3>
<p>If the online databases are not responding due to technical issues or the configuration page doesn’t function as expected, you can manually install the required data. We have provided a GitHub repository with a pre-configured database.</p>
<p>To use this alternative:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Access the repository <a class="reference external" href="https://github.com/lorenzocasbarra/pathlay-dbs">here</a>.</p></li>
<li><p>Download the pathlay_data folder.</p></li>
<li><p>Overwrite the pathlay_data folder on your system with the one from the repository.</p></li>
</ol>
</div></blockquote>
</section>
</section>
<section id="home-manager">
<span id="usage-home"></span><h2>Home Manager<a class="headerlink" href="#home-manager" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
</div>
<p>The Home section is accessible by clicking the “Home” icon from the Welcome page.
In the Home Page you can create, load and modify an Experiment Package with a more intuitive approach.</p>
<div class="toctree-wrapper compound" id="usage-home-create">
</div>
<section id="adding-a-new-experiment-package">
<span id="usage-create-add"></span><h3>Adding a new Experiment Package<a class="headerlink" href="#adding-a-new-experiment-package" title="Link to this heading"></a></h3>
<p>To create an Experiment Package from scratch the first thing to do is to click the “Add New” button.</p>
</section>
<section id="uploading-a-new-experiment-package">
<span id="usage-home-create-upload"></span><h3>Uploading a new Experiment Package<a class="headerlink" href="#uploading-a-new-experiment-package" title="Link to this heading"></a></h3>
<p>To upload an Experiment Package the first thing to do is to click the “Upload” button.</p>
<section id="editing-an-experiment-package">
<span id="usage-home-edit"></span><h4>Editing an Experiment Package<a class="headerlink" href="#editing-an-experiment-package" title="Link to this heading"></a></h4>
<div class="toctree-wrapper compound">
</div>
<p>To edit the configuration of an Experiment Package, click on the “Edit Pack” button on the package’s blue box in the Home section.
You will be redirected to the configuration page.</p>
<p>The configuration articulates in three sections, explained in detail in the following sections:</p>
<ol class="arabic simple">
<li><p>General information editing and reference organism assigning</p></li>
<li><p>Dataset configuration</p></li>
<li><p>Ontologies selection</p></li>
</ol>
<p>After the editing is complete you can save the configuration by clicking the “Save” button.</p>
</section>
</section>
<section id="edit-experiment-s-general-information">
<span id="usage-home-edit-generic"></span><h3>Edit Experiment’s General Information<a class="headerlink" href="#edit-experiment-s-general-information" title="Link to this heading"></a></h3>
<p>By editing the first section in the configuration page you can perform the following:</p>
<ol class="arabic simple">
<li><p>Assign a title to the package</p></li>
<li><p>Assign a reference organism to the package by selecting it from the “Organism” selector.</p></li>
<li><p>Write a comment</p></li>
</ol>
</section>
<section id="configure-a-dataset">
<span id="usage-home-edit-dataset"></span><h3>Configure a Dataset<a class="headerlink" href="#configure-a-dataset" title="Link to this heading"></a></h3>
<p>The configuration of a dataset requires the following general steps:</p>
<ol class="arabic simple">
<li><p>Select what -omic science is representative of the dataset by using the “Select omic” selector</p></li>
<li><p>Select what kind of ID is being adopted in your dataset</p></li>
<li><p>Select which column is holding the IDs</p></li>
<li><p>Select which column is holding an Effect Size Static value (e.g. logFC), if any</p></li>
<li><p>Select which column is holding a p-value, if any</p></li>
<li><p>Upload your dataset in a tab-separated text format by either copy and paste it in the coloured text area, or by uploading it using the “Upload {omic} pack” button below</p></li>
</ol>
</section>
<section id="gene-ontologies">
<span id="usage-home-edit-ont"></span><h3>Gene Ontologies<a class="headerlink" href="#gene-ontologies" title="Link to this heading"></a></h3>
<p>Your Home Page also provides a list of cellular compartments that are nothing less than Gene Ontologies.
Each one of the entries represents a gene list of Entrez Gene IDs and you can select the ones you are more interested in by clicking the “+” button located on the right side of each box.
A search box is also provided to filter the ones you are most interested in.
Gene Ontologies of interest will be added to the other list alongside the preselected ones: Nucleus, Cytoplasm, Mitochondrion, Lysosome, Endoplasmic Reticulum, Golgi Apparatus, Cytoskeleton, Membrane.
You can of course remove them by clicking the “-” button in the selected boxes.
Cellular compartments selected will be found in a dropdown menu after PathLay has completed the analysis and it’s ready to show the results on functional pathways.
Once a cellular compartment is chosen, only genes and proteins localized in that compartment will remain displayed on the pathways (see more “Pathway Explorer”).</p>
<section id="downloading-an-experiment-package">
<span id="usage-home-download"></span><h4>Downloading an Experiment Package<a class="headerlink" href="#downloading-an-experiment-package" title="Link to this heading"></a></h4>
<p>The “Download” button allows you to package the currently loaded datasets and configurations in a zip archive and export it from PathLay for other uses, for instance, this archive can be uploaded as a new Experiment Package into another PathLay account.</p>
</section>
</section>
</section>
<section id="configuration-page">
<span id="usage-access"></span><h2>Configuration Page<a class="headerlink" href="#configuration-page" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
</div>
<p>The Configuration Page is accessible by selecting “Analysis Configurator” in the Welcom page and then clicking the “Configure Analysis” button for the desired Experiment Package.
If it isn’t the first time you do this, you can also skip the configuration step and run PathLay with the last configuration available by clicking “Run Last Configuration”.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/access.png"><img alt="access" src="../_images/access.png" style="width: 100%;" /></a>
</figure>
<p id="usage-access-mapdb">The “Select Maps DB” selector allows you to choose which database will be used to load pathways from.</p>
<figure class="align-center" id="id3">
<a class="reference internal image-reference" href="usage/img/general_conf_sel1.png"><img alt="Data selection interface" src="usage/img/general_conf_sel1.png" style="width: 100%;" /></a>
<figcaption>
<p><span class="caption-text">Screenshot of the “Select Maps DB” interface in the data selection step.</span><a class="headerlink" href="#id3" title="Link to this image"></a></p>
</figcaption>
</figure>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Databases supported may vary depending on the reference organism selected, as it is now PathLay supports only Homo sapiens and Mus musculus which are both supported by KEGG and WikiPathways.</p>
</div>
<p id="usage-access-data">The “Select Datasets to be used” selector allows you to select the datasets you want to use by simply check the boxes you are interested in between “Genes”, “Proteins”, “miRNAs”, “Methylations”, “Chromatin Status” and “Metabolites”.</p>
<figure class="align-center" id="id4">
<a class="reference internal image-reference" href="usage/img/general_conf_sel2.png"><img alt="Data selection interface" src="usage/img/general_conf_sel2.png" style="width: 100%;" /></a>
<figcaption>
<p><span class="caption-text">Screenshot of the “Select Datasets to be used” interface in the data selection step.</span><a class="headerlink" href="#id4" title="Link to this image"></a></p>
</figcaption>
</figure>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Whenever a checkbox is disabled, it means that either the dataset was not found available or that the setup in the Home page was not performed accordingly (i.e. the content of the columns was not pointed out correctly).</p>
</div>
<div class="toctree-wrapper compound" id="usage-data-types">
</div>
<p>Using the “Select omic” selector you can switch between the filtering configurations available for each dataset.
Some of these configurations are common to every data type, while others, related to the Transcription Factors (TFs) and the Non Differentially Expressed components (NoDE), are peculiar to a few of them.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/filters_gene.png"><img alt="Data selection interface" src="../_images/filters_gene.png" style="width: 50%;" /></a>
</figure>
<section id="filtering-options">
<span id="usage-data-filters"></span><h3>Filtering Options<a class="headerlink" href="#filtering-options" title="Link to this heading"></a></h3>
<p>There are three filters available for each data type, which can be enabled by checking the respective box and requires a threshold value to be written in the input field next to it. These three filters are summarized below:</p>
<blockquote>
<div><ul class="simple">
<li><p>Effect Size “&lt;” : filters out all the components with an Effect Size value greater than the threshold selected</p></li>
<li><p>Effect Size “&gt;” : filters out all the components with an Effect Size value smaller than the threshold selected</p></li>
<li><p>p-value “&lt;” : filters out all the components with a p-value value greater than the threshold</p></li>
</ul>
</div></blockquote>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/filters_gene_sel2.png"><img alt="access" src="../_images/filters_gene_sel2.png" style="width: 50%;" /></a>
</figure>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Whenever a threshold field displays a red background, it means that the aforementioned threshold is not valid and contains a typing error. In this situation if the related filter checkbox remains enabled, the submit button will disappear until this error is either fixed by changing the threshold value or by disabling the filter.</p>
</div>
</section>
<section id="usage-data-id-only">
<span id="id-preservation"></span><h3>ID Preservation<a class="headerlink" href="#usage-data-id-only" title="Link to this heading"></a></h3>
<p>ID Preservation can be enabled for every data type by checking the “Preserve non DE IDs” checkbox.
This feature will not let PathLay to discard IDs found in the dataset without an Effect Size value.
These “ID Only” components will be represented with a different palette of colors since they do not provide any information regarding their differential expression, but they will be still integrated with the other datasets.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/filters_gene_sel2.png"><img alt="access" src="../_images/filters_gene_sel2.png" style="width: 50%;" /></a>
</figure>
</section>
<section id="transcription-factors">
<span id="usage-access-tfs"></span><h3>Transcription Factors<a class="headerlink" href="#transcription-factors" title="Link to this heading"></a></h3>
<div class="toctree-wrapper compound">
</div>
<p>PathLay also supports GTRD as a database and can display transcription factors (TFs) that interacts with genes coming from a Transcriptomic or Proteomic dataset and as well recognize if those gene/protein IDs actually refers directly to a transcription factor.
Transcription factors will be loaded and linked to the gene IDs in the Transcriptomic dataset if the “Enable TFs” feature is enabled. The same goes for the proteomic dataset which has its own “Enable TFs” feature.
If a transcription factor is found in the Transcriptomic dataset it will be linked to any related gene also present in the dataset.
Whenever Transcriptomic and Proteomic datasets are enabled together and the “Enable TFs” feature is enabled for at least one of them, once a TF is recognized, its targets will be looked for both in the Transcriptomic and Proteomic dataset to guarantee an high level degree of integration.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/filters_gene_sel3.png"><img alt="access" src="../_images/filters_gene_sel3.png" style="width: 50%;" /></a>
</figure>
<p id="usage-access-id-only-tfs">Normally, if a Transcription Factor is recognized in an ID Only component found in the Transcriptomic or Proteomic dataset (when of course the “Preserve non DE IDs” feature is enabled for them), it is just loaded as a DE component and not represented as a TF.
To overcome this, one should enable the “Preserve IDs for TFs” feature in the Gene or Protein configuration depending on the scope of the analysis.
Transcription Factors recognized in this manner will be represented as small oragne squares on the right side of the main indicators.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/filters_gene_sel4.png"><img alt="access" src="../_images/filters_gene_sel4.png" style="width: 50%;" /></a>
</figure>
</section>
<section id="non-differentially-expressed-ids">
<span id="usage-access-nodeg"></span><h3>Non Differentially Expressed IDs<a class="headerlink" href="#non-differentially-expressed-ids" title="Link to this heading"></a></h3>
<div class="toctree-wrapper compound">
</div>
<p>Whenever a miRNa, Methylation or Chromatin status dataset is enabled, miRNas, methylations and chromatin statuses will be represented alongside the related genes.
PathLays’s approach is to integrate the information held within different datasets whenever is possible, hence when a Transcriptomic or Proteomic dataset is provided and enabled alongside the afromentioned ones, the standard procedure adopted is to represent an integration of these informations using genes and proteins provided as a scaffold on which miRNAs, methylations and chromatin statuses impact.
Anyway It is possible to represent miRNas, methylations or chromatin statuses despite the absence of their related genes in the Transcriptomic dataset, by checking the “No DE Loading” feature in their respective configurations.
This feature allows PathLay to display miRnas, methylations and chromatin statuses on grey indicators that represent a non-differentially expressed gene (i.e. a gene not provided in other datasets).
This feature specifically works with the IDs of miRNAs, methylations and chromatin statuses linked to an Effect Size value, while the IDs without an Effect Size value will be only linked and represented alongside those genes or proteins that are effectively provided by their respectiv datasets, unless the “No DE Loading From Preserved IDs” feature is enabled.
By doing this, ID only components coming from these thre datasets will also be able to call out Non DE genes and proteins.
Graphically speaking this is translated into grey indicators surrounded by smaller circles and squares coloured in orange.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/filters_urna_sel1.png"><img alt="access" src="../_images/filters_urna_sel1.png" style="width: 50%;" /></a>
</figure>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/filters_urna_sel2.png"><img alt="access" src="../_images/filters_urna_sel2.png" style="width: 50%;" /></a>
</figure>
<p id="usage-access-no-de-tfs">The “Load Non DE from TFs” feature provides the possibility to let any differentially expressed TF found to be linked to genes and proteins not provided in the datasets, in a similar fashion of the “No DE Loading” feature seen in the miRNAs, Methylations and Chromatin Statuses configurations.
As previously explained, ID only components will not be allowed to call out for Non Differentially Expressed Genes or Proteins unless this feature is manually enabled and TFs make no exception in this regard: enabling the “Load Non DE from Preserved TFs” feature will do the job.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/filters_gene_sel5.png"><img alt="access" src="../_images/filters_gene_sel5.png" style="width: 50%;" /></a>
</figure>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/filters_gene_sel6.png"><img alt="access" src="../_images/filters_gene_sel6.png" style="width: 50%;" /></a>
</figure>
<section id="maps-restriction-procedure">
<span id="usage-access-stat"></span><h4>Maps Restriction Procedure<a class="headerlink" href="#maps-restriction-procedure" title="Link to this heading"></a></h4>
<div class="toctree-wrapper compound">
</div>
<p>It is possible to perform a selection of maps to display in the PathLay - MAPS page by enabling Fisher’s Exact Test for desired datasets.
Maps are selected with a statistical approach based on the Fisher’s Exact Test and since PathLay integrates different datasets, it is also possible to select which dataset will contribute on the map selection.
By default the test is performed on each dataset enabled and it will produce a list of maps considered of importance, all the lists produced are then joined into one.
There are two tweaks available that can change the result of this step: the Intersect option and the Pooling option.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/stat.png"><img alt="access" src="../_images/stat.png" style="width: 50%;" /></a>
</figure>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>To perform the test you need to provide a dataset that holds a p-value column.</p>
</div>
</section>
</section>
<section id="pooling">
<h3>Pooling<a class="headerlink" href="#pooling" title="Link to this heading"></a></h3>
<p>Eabling the Pooling option will affect how the Transcriptomic, Proteomic, miRNomic, Methylomic and Cromatin Status datasets are treated when enabled for the test.
When enabled, the IDs of the above datasets will be merged and the test is performed on that single list.</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/stat_sel1.png"><img alt="access" src="../_images/stat_sel1.png" style="width: 40%;" /></a>
</figure>
</section>
<section id="intersect">
<h3>Intersect<a class="headerlink" href="#intersect" title="Link to this heading"></a></h3>
<p>Enabling the Intersect option will affect which maps will be selected after the test has been performed.
By default map lists are joined into one, meaning that even if a map has been found relevant in relation to just one of the datasets, it will be kept and indicators will be plotted on it.
With the Intersect feature enabled, only maps that are considered relevant for all the datasets enabled are kept (i.e. the ones that appear in every map list coming from the tests performed).</p>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/stat_sel2.png"><img alt="access" src="../_images/stat_sel2.png" style="width: 40%;" /></a>
</figure>
</section>
</section>
<section id="pathway-explorer">
<span id="usage-results"></span><h2>Pathway Explorer<a class="headerlink" href="#pathway-explorer" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
</div>
<p>The “PathLay - MAPS” page will be displayed after the configuration and the submitting of an analysis on an Experiment Package in your home. IDs coming from your datasets will be portrayed as different indicators on pathways depending on the Analysis Mode chosen (see more “Selecting an Analysis Mode”). This page allows you to:</p>
<blockquote>
<div><ul class="simple">
<li><p>Display your datasets on pathways</p></li>
<li><p>Filter pathways by selecting a specific gene/protein/miRNA/TF ID or by multiple IDs</p></li>
<li><p>Select a cellular compartment and display only the IDs localized in it</p></li>
<li><p>Select indicators and display their infos on the clipboard</p></li>
<li><p>Save indicators of interest and download them as a tab-separated text file</p></li>
<li><p>Filter pathways by selecting an “Agreement” between Gene/Proteins and Methylation, miRNAs or TFs</p></li>
</ul>
</div></blockquote>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The “PathLay - MAPS” page is directly accessible after you hit the “Submit” button allowing you to navigate through pathways right away but depending on how much data PathLay needs to process, some indicators may not be displayed correctly from the go and require more time.</p>
</div>
<p id="usage-results-mapsel">Pathways available are listed in alphabetical order in the dropdown menu on the top left corner. The pathway name and ID are displayed and the numbers between round parenthesis give you information on how many indicators are loaded on the pathway. Selecting one pathway from this menu will instantly display it with its the indicators.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The pathway selector will display all the pathways available only when no other selections have been performed. Whenever an ID/Agreement/Cellular compartment based selection is performed, this dropdown menu will provide only those pathways that satisfies the parameters of the selection. To return at the initial, default state, simply reset your selections.</p>
</div>
<figure class="align-center" id="usage-results-legend">
<a class="reference internal image-reference" href="../_images/legend.png"><img alt="legend" src="../_images/legend.png" style="width: 100%;" /></a>
</figure>
<figure class="align-center">
<a class="reference internal image-reference" href="../_images/legend2.png"><img alt="legend" src="../_images/legend2.png" style="width: 100%;" /></a>
</figure>
<p id="usage-results-menu">The “PathLay - MAPS” page provides different features in six sections accessible from a menu located in the upper right corner of the page:</p>
<blockquote>
<div><ul class="simple">
<li><p>Highlight: Provides the selectors to highlight indicators containing specific components</p></li>
<li><p>Screenshot: Creates a downloadable image of a map with its indicators</p></li>
<li><p>Clipboard: Open or close Clipboard</p></li>
<li><p>Settings: Allows tuning of indicators transparency and size</p></li>
<li><p>Selectors: Provides the main query system for pathway and indicator filtering</p></li>
<li><p>?: Displays helpful information regarding the interpratation of the indicators</p></li>
</ul>
</div></blockquote>
<p>These sections and the respective features will be explained in detail later.</p>
<p id="usage-results-highlight">In this sections two selectors are displayed and their content varies depending on the features enabled during configuration.
The first selector lists the categories of components available between: Genes, Proteins, TFs, miRNAs, Methylations, Chromatin Statuses and Metabolites.
Selecting one data type will switch to the respective component selector filled with the available IDs.
When an ID is chosen, all the indicators related to it will be highlighted with a pink dotted border, and the maps not containing any indicator of interest will be disabled from the map selector.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To reset the highlight simply select the first option from the ID selector.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The highlight feature will highlight only those indicators currently available from the displayable maps.
This means that when a query from the “Selectors” section has been performed, the selectors in the Highlight section will be filled only with the components represented on the maps that remained available after it.</p>
</div>
<p id="usage-results-clipboard">The Clipboard is a freely movable window that will be displayed as soon as the “Clipboard” writing is clicked in the menu.
It is composed by two boxes, the “Info” box and the “Selected” box.
The “Info” box will display the information related to a clicked indicator, this includes every ID that the indicator is portaying, with their differential expression values and other IDs eventually linked.
Cliking on the IDs in the “Info” box will open up a page in the online database of reference for that data type:</p>
<blockquote>
<div><ul class="simple">
<li><p>NCBI for genes</p></li>
<li><p>UniProtKB for proteins</p></li>
<li><p>Mirtarbase for miRNAs</p></li>
<li><p>gsea-msigdb for TFs.</p></li>
</ul>
</div></blockquote>
<p>Once you have clicked an indicator on a pathway and its information is displayed on the Clipboard, you can click the “Add” button below the “Info” box and save the indicator inside the Clipboard’s “Selected” box.
It will be listed as a smaller indicator under the name of the pathway from where the selction has been performed.
Selecting this smaller indicator in the box will also display its information in the “Info” box and if the “Remove” button is clicked, the indicator will be removed from the “Selected” box.
The “Select” button under the “Selected” box on the Clipboard performs the same operations as the highlight button in the Highlight Section, thus applying a dotted magenta border on the indicators of interest and disabling maps not containing them in the map selector.
The last feature provided by the Clipboard is the “Download” button which allows to download a tab separated text document listing all the information regarding selected indicators currently populating the “Selected” box.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Closing the Clipboard by clicking the “X” on the top right corner will not delete any selected indicator from the “Selected” box.</p>
</div>
<div class="toctree-wrapper compound" id="usage-results-logical">
</div>
<dl class="simple">
<dt>This section contains the main query system of the “PathLay - MAPS” page allowing to perform a more sophisticate three level selection by:</dt><dd><ul class="simple">
<li><p>IDs</p></li>
<li><p>Agreement between two components</p></li>
<li><p>Ontology (Cellular Compartments)</p></li>
</ul>
</dd>
</dl>
<p>These selections can be performed by adding a query tag using one of the three selectors and clicking the respective “Add” button.
Once all the query tags you are interested in are added to the pool on the right, the query can be performed by clicking the “Select” button.</p>
<dl class="simple">
<dt>A brief summary of the rationale behind the query is the following:</dt><dd><ul class="simple">
<li><p>The IDs provided are rounded up and only those maps that contains indicators representing them are allowed to proceed in the query. Is important to note that this step works with an “AND” approach so, whenever a map does not represent even one of the IDs provided, it is discarded and thus removed from the selector.</p></li>
<li><p>The Agreements provided are rounded up and the indicators that are currently displayable or that passed the previous step of the query, are checked for all the Agreements. Whenever an agreement is not satisfied, the indicator is hidden and if all the indicators in a specific map don’t satisfy the agreements, the map is discarded and removed from the selector.</p></li>
<li><p>The Ontologies provided are rounded up and the indicators that are currently displayable or that passed the previous step of the query, are checked for components that refers to the cellular compartments selected. Whenever an indicator doesn’t refer to one of the compartments selected, it is hidden and if all the indicators in a specific map don’t refer to the cellular compartments, the map is discarded and removed from the map selector.</p></li>
</ul>
</dd>
</dl>
<p>To reset the interface to its default state simply click the “Reset” button next to it.
Details on each type of query tag are provided below.</p>
<section id="select-by-ids">
<span id="usage-results-logical-byids"></span><h3>Select by IDs<a class="headerlink" href="#select-by-ids" title="Link to this heading"></a></h3>
<p>The first row in the Logical Section is composed by a selector that allows the selection of a data type between “Gene”, “Protein”, TFs, “miRNA”, “Methylation”, “Chromatin” and “Metabolite”, depending on your analysis configuration.
Selecting one data type will pop up another selector with a list of IDs of the data type chosen, in a similar fashion of the HighLight section.
Once an ID is selected, it can be added to the query pool “IDs Selected”.
Every ID added to the pool will be represented as a query tag and can be removed from the pool by clicking the “X” button on the right.
IDs’ query tag have colors are assigned considering the data type:</p>
<blockquote>
<div><ul class="simple">
<li><p>Light blue for genes</p></li>
<li><p>Purple for proteins</p></li>
<li><p>Red for miRNAs</p></li>
<li><p>Yellow for Metabolites</p></li>
<li><p>Grey for TFs, Methylations and Chromatin Statuses</p></li>
</ul>
</div></blockquote>
<p>The “Select” button will perform a pathway selection using the pool content as criteria: considering a strictly ID based query, only those pathways portraying all the IDs in the pool will be displayed in the pathway selector and kept available.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It is possible that loaded pathways cannot satisfy the requirements for the selection made. In this case the pathway selector will become empty. Resetting the selection with the “Reset” button will restore the pathway selector at its default state.</p>
</div>
</section>
<section id="select-by-agreement">
<span id="usage-results-logical-byagreement"></span><h3>Select by Agreement<a class="headerlink" href="#select-by-agreement" title="Link to this heading"></a></h3>
<p>The second row in the Logical Section is composed by a series of selectors that allows the setup of the so called “Agreement” which is a feature that lets you select two related data types and select what type of correlation you are interested in. The main data type can be selected with the first selector which provides a choice between “Gene” and “Protein”, and the second data type can be selected with the other selector which allows a choice between “Gene” or “Protein” (depending on your main data type of choice), “miRNA”, “Methylation” or “TF”. After the two data types are selected, the last step to do is to choose a “Positive Agreement” or a “Negative Agreement” with the last selector and hit the “Add” button. The configured agreement will be displayed in the “Agreements Selected” pool positioned under the “IDs Selected” pool. Once the select button is clicked, indicators that don’t satisfy the requirements of the agreement will be hidden, thus if a pathway doesn’t have visible indicators it will not be included in the pathway selector. Multiple agreements can be added to the pool to make the selection more stringent. How various agreements work is reported in the following table:</p>
<p>As an example, let’s suppose we want to examine only the indicators including a gene and a miRNA in a positive agreement: it is only necessary to select “Gene”, “miRNA” and “Positive” from the selectors provided in the Logical section. A positive agreement between a gene and a miRNA is based on a negative correlation, so the indicators highlighted will either have an up-regulated gene and a down-regulated miRNA or a down-regulated gene and an up-regulated miRNA. As a next step, we could also be looking at methylations: easily enough, agreement selections can be stacked so we could also select “Gene”, “Methylation”, “Positive” and highlight only those indicators that satisfy both the agreements.
As before, to remove it the “X” button on the right or the “Reset” button must be clicked.</p>
<section id="settings">
<span id="usage-results-settings"></span><h4>Settings<a class="headerlink" href="#settings" title="Link to this heading"></a></h4>
<div class="toctree-wrapper compound">
</div>
<p>The settings section in the “PathLay - MAPS” page will be displayed when clicking the “Show Settings” button and allows to change some graphical parameters for the indicators like size and transparency values.</p>
</section>
</section>
<section id="transparency-and-size">
<span id="usage-results-settings-transparencyandsize"></span><h3>Transparency and Size<a class="headerlink" href="#transparency-and-size" title="Link to this heading"></a></h3>
<p>Transparency and size of the indicators can be increased and decreased with the “Transparency Up”,”Size Up” and “Transparency Down”,”Size Down” buttons respectively.</p>
</section>
<section id="screenshot">
<span id="usage-results-settings-screenshot"></span><h3>Screenshot<a class="headerlink" href="#screenshot" title="Link to this heading"></a></h3>
<p>It is possible to take a shot of the currently displayed pathway by clicking the “Open as image (to save..)” button.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Depending on your browser compatibility this feature can behave differently.</p>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../howto/howto_localserver.html" class="btn btn-neutral float-left" title="Run PathLay as local server with Docker (localhost)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="account/usage_account_index.html" class="btn btn-neutral float-right" title="Account Managing" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Matteo Ramazzotti Lorenzo Casbarra.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-XXXXXXXXXX"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-XXXXXXXXXX', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>