<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage &mdash; PathLay 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> PathLay
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Usage</a><ul>
<li><a class="reference internal" href="#account-managing">Account Managing</a><ul>
<li><a class="reference internal" href="#register-a-new-account">Register a New Account</a></li>
<li><a class="reference internal" href="#login-to-an-existing-account">Login to an Existing Account</a></li>
</ul>
</li>
<li><a class="reference internal" href="#experiment-package">Experiment Package</a><ul>
<li><a class="reference internal" href="#configuration-file">Configuration File</a></li>
<li><a class="reference internal" href="#datasets-files">Datasets Files</a></li>
<li><a class="reference internal" href="#supported-ids">Supported IDs</a></li>
</ul>
</li>
<li><a class="reference internal" href="#home-manager">Home Manager</a><ul>
<li><a class="reference internal" href="#creating-an-experiment-package">Creating an Experiment Package</a></li>
<li><a class="reference internal" href="#loading-an-experiment-package">Loading an Experiment Package</a></li>
<li><a class="reference internal" href="#gene-ontologies">Gene Ontologies</a></li>
<li><a class="reference internal" href="#downloading-an-experiment-package">Downloading an Experiment Package</a></li>
</ul>
</li>
<li><a class="reference internal" href="#configuration-page">Configuration Page</a><ul>
<li><a class="reference internal" href="#selecting-an-analysis-mode">Selecting an Analysis Mode</a></li>
<li><a class="reference internal" href="#non-differentially-expressed-ids">Non Differentially Expressed IDs</a></li>
<li><a class="reference internal" href="#transcription-factors">Transcription Factors</a></li>
<li><a class="reference internal" href="#maps-restriction-procedure">Maps Restriction Procedure</a></li>
<li><a class="reference internal" href="#maps-databases">Maps Databases</a></li>
<li><a class="reference internal" href="#data-selection">Data Selection</a></li>
</ul>
</li>
<li><a class="reference internal" href="#pathway-explorer">Pathway Explorer</a><ul>
<li><a class="reference internal" href="#pathway-navigation">Pathway Navigation</a></li>
<li><a class="reference internal" href="#indicators">Indicators</a></li>
<li><a class="reference internal" href="#highlighting-feature">Highlighting Feature</a></li>
<li><a class="reference internal" href="#the-clipboard">The Clipboard</a></li>
<li><a class="reference internal" href="#the-logical-section">The Logical Section</a><ul>
<li><a class="reference internal" href="#select-by-ids">Select by IDs</a></li>
<li><a class="reference internal" href="#select-by-agreement">Select by Agreement</a></li>
</ul>
</li>
<li><a class="reference internal" href="#settings">Settings</a><ul>
<li><a class="reference internal" href="#transparency-and-size">Transparency and Size</a></li>
<li><a class="reference internal" href="#screenshot">Screenshot</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">PathLay</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Usage</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/usage.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this heading"></a></h1>
<section id="account-managing">
<span id="usage-account"></span><h2>Account Managing<a class="headerlink" href="#account-managing" title="Permalink to this heading"></a></h2>
<section id="register-a-new-account">
<span id="usage-account-new"></span><h3>Register a New Account<a class="headerlink" href="#register-a-new-account" title="Permalink to this heading"></a></h3>
</section>
<section id="login-to-an-existing-account">
<span id="usage-account-login"></span><h3>Login to an Existing Account<a class="headerlink" href="#login-to-an-existing-account" title="Permalink to this heading"></a></h3>
</section>
</section>
<section id="experiment-package">
<span id="usage-specs"></span><h2>Experiment Package<a class="headerlink" href="#experiment-package" title="Permalink to this heading"></a></h2>
<p>In the following section, the structure of a zipped Experiment Package will be explained. Please note that each and avery file listed below can be created and modified from your Home Page direclty, with no need to manually adjust them from outside PathLay.
Each Experiment Package is defined from up to 7 different text files with different extensions:</p>
<table class="colwidths-given docutils align-default" id="id1">
<caption><span class="caption-text">Experiment Package Content</span><a class="headerlink" href="#id1" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 30%" />
<col style="width: 70%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>File Extension</p></th>
<th class="head"><p>Content</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>.conf</p></td>
<td><p>Holds the main configuration of the Experiment Package</p></td>
</tr>
<tr class="row-odd"><td><p>.mrna</p></td>
<td><p>Holds the Transcriptomic dataset</p></td>
</tr>
<tr class="row-even"><td><p>.prot</p></td>
<td><p>Holds the Proteomic dataset</p></td>
</tr>
<tr class="row-odd"><td><p>.mirna</p></td>
<td><p>Holds the miRNomic dataset</p></td>
</tr>
<tr class="row-even"><td><p>.meth</p></td>
<td><p>Holds the Methylomic dataset</p></td>
</tr>
<tr class="row-odd"><td><p>.ont</p></td>
<td><p>Holds the Gene Ontologies entries of interest of the Experiment Package</p></td>
</tr>
<tr class="row-even"><td><p>.meta</p></td>
<td><p>Holds the Metabolomic dataset</p></td>
</tr>
</tbody>
</table>
<p>Every experiment created in your Home Page has a name like “exp1,2,3…” automatically assigned. If you want to manually configure an experiment from scratch you have to name your files with this convention in mind (i.e. exp1.conf, exp1.mrna, exp1.meta etc…).</p>
<section id="configuration-file">
<span id="usage-specs-conf"></span><h3>Configuration File<a class="headerlink" href="#configuration-file" title="Permalink to this heading"></a></h3>
<p>The configuration file for the Experiment Package is structured as a series of tag and value pairs, separated by a “=” character, with each line of the file holding one pair.</p>
<table class="colwidths-given docutils align-default" id="id2">
<caption><span class="caption-text">Configuration File Tags</span><a class="headerlink" href="#id2" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 16%" />
<col style="width: 53%" />
<col style="width: 32%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Tag</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Content</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>expname</p></td>
<td><p>Holds the name of the Experiment Package</p></td>
<td><p>String of text</p></td>
</tr>
<tr class="row-odd"><td><p>comments</p></td>
<td><p>Holds brief comments for the Experiment Package</p></td>
<td><p>String of text</p></td>
</tr>
<tr class="row-even"><td><p>idcol</p></td>
<td><p>Holds the number representing the column with the IDs in the Transcriptomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>datacol</p></td>
<td><p>Holds the number representing the column with the differential expression values in the Transcriptomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>datatype</p></td>
<td><p>Holds the type of differential expression value used in the Transcriptomic dataset</p></td>
<td><p>“logFC” or “FC” or “Raw”</p></td>
</tr>
<tr class="row-odd"><td><p>datadir</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>datathr</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>pcol</p></td>
<td><p>Holds the number representing the column with the p-values in the Transcriptomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>pthr</p></td>
<td><p>Holds the number representing the threshold for the p-values in the Transcriptomic dataset</p></td>
<td><p>Any Number</p></td>
</tr>
<tr class="row-odd"><td><p>midcol</p></td>
<td><p>Holds the number representing the column with the IDs in the miRNomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>mdatacol</p></td>
<td><p>Holds the number representing the column with the differential expression values in the miRNomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>mdatatype</p></td>
<td><p>Holds the type of differential expression value used in the miRNomic dataset</p></td>
<td><p>“logFC” or “FC” or “Raw”</p></td>
</tr>
<tr class="row-even"><td><p>mdatadir</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>mdatathr</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>mpcol</p></td>
<td><p>Holds the number representing the column with the p-values in the miRNomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>mpthr</p></td>
<td><p>Holds the number representing the threshold for the p-values in the miRNomic dataset</p></td>
<td><p>Any Number</p></td>
</tr>
<tr class="row-even"><td><p>metidcol</p></td>
<td><p>Holds the number representing the column with the IDs in the Metabolomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>metdatacol</p></td>
<td><p>Holds the number representing the column with the expression values in the Metabolomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>metdatatype</p></td>
<td><p>Holds the type of differential expression value used in the Metabolomic dataset</p></td>
<td><p>“logFC” or “FC” or “Raw”</p></td>
</tr>
<tr class="row-odd"><td><p>metdatadir</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>metdatathr</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>metpcol</p></td>
<td><p>Holds the number representing the column with the p-values in the Metabolomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>metpthr</p></td>
<td><p>Holds the number representing the threshold for the p-values in the Metabolomic dataset</p></td>
<td><p>Any Number</p></td>
</tr>
<tr class="row-odd"><td><p>methidcol</p></td>
<td><p>Holds the number representing the column with the IDs in the Methylomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>methdatacol</p></td>
<td><p>Holds the number representing the column with the differential expression values in the Methylomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>methdatatype</p></td>
<td><p>Holds the type of differential expression value used in the Methylomic dataset</p></td>
<td><p>“logFC” or “FC” or “Raw”</p></td>
</tr>
<tr class="row-even"><td><p>methdatadir</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>methdatathr</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>methpcol</p></td>
<td><p>Holds the number representing the column with the p-values in the Methylomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>methpthr</p></td>
<td><p>Holds the number representing the threshold for the p-values in the Methylomic dataset</p></td>
<td><p>Any Number</p></td>
</tr>
<tr class="row-even"><td><p>protidcol</p></td>
<td><p>Holds the number representing the column with the IDs in the Proteomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-odd"><td><p>protdatacol</p></td>
<td><p>Holds the number representing the column with the differential expression values in the Proteomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>protdatatype</p></td>
<td><p>Holds the type of differential expression value used in the Proteomic dataset</p></td>
<td><p>“logFC” or “FC” or “Raw”</p></td>
</tr>
<tr class="row-odd"><td><p>protdatadir</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>protdatathr</p></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>protpcol</p></td>
<td><p>Holds the number representing the column with the p-values in the Proteomic dataset</p></td>
<td><p>Integer Number</p></td>
</tr>
<tr class="row-even"><td><p>protpthr</p></td>
<td><p>Holds the number representing the threshold for the p-values in the Proteomic dataset</p></td>
<td><p>Any Number</p></td>
</tr>
</tbody>
</table>
<p>An example of a fully configured .conf file is displayed below:</p>
<div class="literal-block-wrapper docutils container" id="id3">
<div class="code-block-caption"><span class="caption-text">Structure of a .conf file of an Experiment Package with Transcriptomic, Proteomic, miRNomic, Methylomic and Metabolomic datasets associated</span><a class="headerlink" href="#id3" title="Permalink to this code"></a></div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">expname</span><span class="o">=</span><span class="n">LTED</span> <span class="n">vs</span> <span class="n">MCF7</span><span class="o">+</span>
<span class="n">comments</span><span class="o">=</span><span class="n">ANOVA</span> <span class="n">adj</span><span class="o">.</span><span class="n">p</span> <span class="o">+</span> <span class="n">Tukey</span>
<span class="n">idcol</span><span class="o">=</span><span class="mi">8</span>
<span class="n">datacol</span><span class="o">=</span><span class="mi">6</span>
<span class="n">datatype</span><span class="o">=</span><span class="n">logFC</span>
<span class="n">datadir</span><span class="o">=</span><span class="n">out</span>
<span class="n">datathr</span><span class="o">=</span><span class="mi">1</span>
<span class="n">pcol</span><span class="o">=</span><span class="mi">5</span>
<span class="n">pthr</span><span class="o">=</span><span class="mf">0.05</span>
<span class="n">midcol</span><span class="o">=</span><span class="mi">1</span>
<span class="n">mdatacol</span><span class="o">=</span><span class="mi">6</span>
<span class="n">mdatatype</span><span class="o">=</span><span class="n">logFC</span>
<span class="n">mdatadir</span><span class="o">=</span><span class="n">out</span>
<span class="n">mdatathr</span><span class="o">=</span><span class="mi">1</span>
<span class="n">mpcol</span><span class="o">=</span><span class="mi">5</span>
<span class="n">mpthr</span><span class="o">=</span><span class="mf">0.05</span>
<span class="n">metidcol</span><span class="o">=</span><span class="mi">8</span>
<span class="n">metdatacol</span><span class="o">=</span><span class="mi">6</span>
<span class="n">metdatatype</span><span class="o">=</span><span class="n">logFC</span>
<span class="n">metdatadir</span><span class="o">=</span><span class="n">out</span>
<span class="n">metdatathr</span><span class="o">=</span><span class="mi">1</span>
<span class="n">metpcol</span><span class="o">=</span><span class="mi">5</span>
<span class="n">metpthr</span><span class="o">=</span><span class="mf">0.05</span>
<span class="n">methidcol</span><span class="o">=</span><span class="mi">8</span>
<span class="n">methdatacol</span><span class="o">=</span><span class="mi">6</span>
<span class="n">methdatatype</span><span class="o">=</span><span class="n">logFC</span>
<span class="n">methdatadir</span><span class="o">=</span><span class="n">out</span>
<span class="n">methdatathr</span><span class="o">=</span><span class="mi">1</span>
<span class="n">methpcol</span><span class="o">=</span><span class="mi">5</span>
<span class="n">methpthr</span><span class="o">=</span><span class="mf">0.05</span>
<span class="n">protidcol</span><span class="o">=</span><span class="mi">1</span>
<span class="n">protdatacol</span><span class="o">=</span><span class="mi">4</span>
<span class="n">protdatatype</span><span class="o">=</span><span class="n">logFC</span>
<span class="n">protdatadir</span><span class="o">=</span><span class="n">out</span>
<span class="n">protdatathr</span><span class="o">=</span><span class="mi">1</span>
<span class="n">protpcol</span><span class="o">=</span><span class="mi">3</span>
<span class="n">protpthr</span><span class="o">=</span><span class="mf">0.05</span>
</pre></div>
</div>
</div>
<p>The experiment Package is named “LTED vs MCF7+” and we can see it has all the -omics datasets currently supported by PathLay associated to it, and we can read their configurations as it follows:</p>
<ul class="simple">
<li><p>The Transcriptomic datataset has its IDs stored in the 8th column, logFC values stored in the 6th column and p-values stored in the 5th column. IDs with a p-value minor than 0.05 and logFC minor or major than 1 will be kept during PathLay’s analysis.</p></li>
<li><p>The miRNomic datataset has its IDs stored in the 1st column, logFC values stored in the 6th column and p-values stored in the 5th column. IDs with a p-value minor than 0.05 and logFC minor or major than 1 will be kept during PathLay’s analysis.</p></li>
<li><p>The Metabolomic datataset has its IDs stored in the 8th column, logFC values stored in the 6th column and p-values stored in the 5th column. IDs with a p-value minor than 0.05 and logFC minor or major than 1 will be kept during PathLay’s analysis.</p></li>
<li><p>The Methylomic datataset has its IDs stored in the 8th column, logFC values stored in the 6th column and p-values stored in the 5th column. IDs with a p-value minor than 0.05 and logFC minor or major than 1 will be kept during PathLay’s analysis.</p></li>
<li><p>The Proteomic datataset has its IDs stored in the 1st column, logFC values stored in the 4th column and p-values stored in the 3rd column. IDs with a p-value minor than 0.05 and logFC minor or major than 1 will be kept during PathLay’s analysis.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Experiment Packages can be created and configured in your Home Page with a more intuitive approach. Once the “Save” button is clicked the .conf file will be automatically generated and saved in your home folder.</p>
</div>
</section>
<section id="datasets-files">
<span id="usage-specs-data"></span><h3>Datasets Files<a class="headerlink" href="#datasets-files" title="Permalink to this heading"></a></h3>
<p>Each dataset file (i.e. .mrna, .prot, .mirna, .meta and .meth files) is a tab separated file that can have any number of columns but at least one, two or three, depending on the Analysis Mode you will choose for PathLay (see more in Configuration Page section). An “ID Only” analysis just requires an ID list in your dataset file, an “ID + DE” analysis requires an ID list with Expression Values associated and a “Full” analysis requires all the above plus a list of p-values associated to each ID.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Datasets can be copied and pasted in their related text areas in your Home Page. Once the “Save” button is clicked the dataset file will be created in your home folder and named after the experiment.</p>
</div>
</section>
<section id="supported-ids">
<span id="usage-specs-ids"></span><h3>Supported IDs<a class="headerlink" href="#supported-ids" title="Permalink to this heading"></a></h3>
<table class="colwidths-given docutils align-default" id="id4">
<caption><span class="caption-text">Compatible IDs to use in your Datasets</span><a class="headerlink" href="#id4" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Dataset Type</p></th>
<th class="head"><p>Supported ID</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Gene</p></td>
<td><p>Entrez Gene ID</p></td>
</tr>
<tr class="row-odd"><td><p>Protein</p></td>
<td><p>UniProtKB Entry</p></td>
</tr>
<tr class="row-even"><td><p>miRNA</p></td>
<td><p>miRNA Name</p></td>
</tr>
<tr class="row-odd"><td><p>Metabolite</p></td>
<td><p>KEGG Compounds</p></td>
</tr>
<tr class="row-even"><td><p>Methylation</p></td>
<td><p>Entrez Gene ID</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="home-manager">
<span id="usage-home"></span><h2>Home Manager<a class="headerlink" href="#home-manager" title="Permalink to this heading"></a></h2>
<p>Once an account is registered the Home Page becomes available, while for an already registered account it is accessible by clicking the “Manage Home” button in the login page. In the Home Page you can create, load and modify an Experiment Package with a more intuitive approach.</p>
<section id="creating-an-experiment-package">
<span id="usage-home-create"></span><h3>Creating an Experiment Package<a class="headerlink" href="#creating-an-experiment-package" title="Permalink to this heading"></a></h3>
<p>To create an Experiment Package from scratch the first thing to do is to click the “Add New” button and then assign it a name in the “Experiment Title” text box, you can also comment it in the “Comments” text box located below.
You can now proceed to load your datasets on PathLay. Data types supported may come from Transcriptomic (Genes), miRNomic (miRNAs), Metabolomic (Metabolites), Methylomic (Methylations) and Proteomic (Proteins) experiments. Each of the aforementioned data types have a coloured and manually resizable text box in which is possible to copy and paste a tab separated dataset (see “Dataset Files”):</p>
<ul class="simple">
<li><p>Genes datasets can be loaded in the light blue text box</p></li>
<li><p>miRNAs datasets can be loaded in the red text box</p></li>
<li><p>Metabolites datasets can be loaded in the yellow text box</p></li>
<li><p>Methylations datasets can be loaded in the green text box</p></li>
<li><p>Proteins datasets can be loaded in the purple text box</p></li>
</ul>
<p>Only one text box is displayed at a time, to change this simply click on the dropdown menu “Data Type” and select one of the five data types.
When a dataset is displayed, its configuration becomes available for you to see and set up.
This setup requires you the following:</p>
<ul class="simple">
<li><p>Mark the column number of the dataset that stores the IDs using the “ID Column” text box</p></li>
<li><p>Select the type of differential expression values stored in the dataset using the dropdown menu, “logFC”, “FC” and “Raw” are your currently available options</p></li>
<li><p>Mark the column number of the dataset that stores the differential expression values using the text box next to the aforementioned dropdown menu</p></li>
<li><p>Select which differential expression values should be kept during the analysis selecting  “&gt;”, “&lt;”, “&lt;&gt;” or “&gt;&lt;” from the dropdown menu and writing a threshold</p></li>
<li><p>Mark the column number of the dataset that stores p-values and write a threshold for them</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Depending on the Analysis Mode you will choose for PathLay, not all of the above configurations must be performed:</p>
<blockquote>
<div><ul class="simple">
<li><p>An “ID Only” analysis just requires an ID list in your dataset file</p></li>
<li><p>an “ID + DE” analysis requires an ID list with Expression Values associated</p></li>
<li><p>A “Full” analysis requires all the above plus a list of p-values associated to each ID.</p></li>
</ul>
</div></blockquote>
</div>
<p>Once configuration is done you can click the “Save” button and the Experiment Package will be created. The “Save” button will of course also save any modification you will submit to PathLay when you will reload this Experiment Package in the future.</p>
</section>
<section id="loading-an-experiment-package">
<span id="usage-home-load"></span><h3>Loading an Experiment Package<a class="headerlink" href="#loading-an-experiment-package" title="Permalink to this heading"></a></h3>
<p>To load an existing Experiment Package from your home folder, select its name from the dropdown menu in the top left corner and hit the “Load” button. Datasets and configurations will be displayed in their relative fields.
If you have a zipped Experiment Package and you want to load it on PathLay, simply click the “Browse” button in the top right corner and select the zipped archive, then click the “Upload” button and load it.</p>
</section>
<section id="gene-ontologies">
<span id="usage-home-ont"></span><h3>Gene Ontologies<a class="headerlink" href="#gene-ontologies" title="Permalink to this heading"></a></h3>
<p>Your Home Page also provides a list of cellular compartments that are nothing less than Gene Ontologies. Each one of the entries represents a gene list of Entrez Gene IDs and you can select the ones you are more interested in by clicking the “+” button located on the right side of each box. Gene Ontologies of interest will be added to the other list alongside the preselected ones: Nucleus, Cytoplasm, Mitochondrion, Lysosome, Endoplasmic Reticulum, Golgi Apparatus, Cytoskeleton, Membrane. You can of course remove them by clicking the “-” button in the selected boxes.
Cellular compartments selected will be found in a dropdown menu after PathLay has completed the analysis and it’s ready to show the results on functional pathways. Once a cellular compartment is chosen, only genes and proteins localized in that compartment will remain displayed on the pathways (see more “Pathway Explorer”).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The selection of Gene Ontologies is as well part of your configuration for the Experiment Package, so any change you provide will be saved with the “Save” button.</p>
</div>
</section>
<section id="downloading-an-experiment-package">
<span id="usage-home-download"></span><h3>Downloading an Experiment Package<a class="headerlink" href="#downloading-an-experiment-package" title="Permalink to this heading"></a></h3>
<p>The “Download” button allows you to package the currently loaded datasets and configurations in a zip archive and export it from PathLay for other uses, for instance, this archive can be uploaded as a new Experiment Package into another PathLay account.
You can also download the entirety of your home folder (i.e. all the datasets and configurations of all the experiments) with the “Download Home” button.</p>
</section>
</section>
<section id="configuration-page">
<span id="usage-access"></span><h2>Configuration Page<a class="headerlink" href="#configuration-page" title="Permalink to this heading"></a></h2>
<p>The Configuration Page is accessible from the Login Page by clicking “Access PathLay” once you input your credentials. From this page you will be able to run PathLay’s analysis on an Experiment Package.
First thing to do is select an Experiment Package and load it using the dropdown menu and the “Load” button.</p>
<section id="selecting-an-analysis-mode">
<span id="usage-access-mode"></span><h3>Selecting an Analysis Mode<a class="headerlink" href="#selecting-an-analysis-mode" title="Permalink to this heading"></a></h3>
<p>You can choose a Mode from the dropdown menu, which provides three options:</p>
<blockquote>
<div><ul class="simple">
<li><p>Full: This mode needs a dataset where IDs, DEs and p-values are provided. It will represents the presence of statistically relevant IDs on pathways with a color scheme indicating their differential expression. Only the IDs with a DE and a p-value compatible with the thresholds configured in the Home Page will be displayed.</p></li>
<li><p>ID Only: This mode will only represent the presence of the IDs on pathways.</p></li>
<li><p>ID + DE: This mode needs a dataset where IDs and DEs are provided. It will represents the presence of IDs on pathways with a color scheme indicating their differential expression. Only the IDs with a DE compatible with the threshold configured in the Home Page will be displayed.</p></li>
</ul>
</div></blockquote>
</section>
<section id="non-differentially-expressed-ids">
<span id="usage-access-nodeg"></span><h3>Non Differentially Expressed IDs<a class="headerlink" href="#non-differentially-expressed-ids" title="Permalink to this heading"></a></h3>
<p>Whenever a miRNomic or a Methylomic dataset is in use, miRNas and methylations will be represented alongside the related genes. PathLays’s approach is to integrate the information held within different datasets whenever is possible, hence when a Transcriptomic dataset is provided together with a miRNomic or/and a Methylomic one, the standard procedure adopted is to represent only miRNas and methylations related to the genes provided in the Transcriptomic dataset. It is possible to represent miRNas or methylations despite them being related to genes in a Transcriptomic dataset, by checking “miRNAs” or/and “Methylations” in the “Load NODEGs from:” section. This feature allows PathLay to display miRnas and methylations as grey indicators that represent a non-differentially expressed gene (i.e. a gene not provided in other datasets).</p>
</section>
<section id="transcription-factors">
<span id="usage-access-tfs"></span><h3>Transcription Factors<a class="headerlink" href="#transcription-factors" title="Permalink to this heading"></a></h3>
<p>PathLay also supports GTRD as a database and can display transcription factors (TFs) that interacts with genes coming from a Transcriptomic dataset and as well recognize if those gene IDs actually refers directly to a transcription factor. Transcription factors contained in GTRD database will be loaded and linked to the gene IDs in the Transcriptomic dataset, if a transcription factor is found in the Transcriptomic dataset it will also be linked to any related gene in the dataset. An option for TFs is present in the “Load NODEGs from:” section alongside miRNAs and methylations, when enabled it allows any transcription factor found in the transcriptomic dataset to be linked to any ID not included in the dataset with the same approach as miRNAs and methylations.</p>
</section>
<section id="maps-restriction-procedure">
<span id="usage-access-stat"></span><h3>Maps Restriction Procedure<a class="headerlink" href="#maps-restriction-procedure" title="Permalink to this heading"></a></h3>
</section>
<section id="maps-databases">
<span id="usage-access-mapdb"></span><h3>Maps Databases<a class="headerlink" href="#maps-databases" title="Permalink to this heading"></a></h3>
<p>The “Maps Database” section allows you to choose which database will be used to load pathways from, it currently provides two possibilities, both for Homo sapiens pathways: KEGG and WikiPathways.</p>
</section>
<section id="data-selection">
<span id="usage-access-data"></span><h3>Data Selection<a class="headerlink" href="#data-selection" title="Permalink to this heading"></a></h3>
<p>The final step before launching the analysis is to select the datasets you want to use in the “Select Maps using” section and simply check the boxes you are interested in between “Genes”, “Proteins”, “miRNAs”, “Methylations” and “Metabolites”. Once it is all set, hit the “Submit” button and wait for PathLay to display the pathways.</p>
</section>
</section>
<section id="pathway-explorer">
<span id="usage-results"></span><h2>Pathway Explorer<a class="headerlink" href="#pathway-explorer" title="Permalink to this heading"></a></h2>
<p>The “Analysis Results” page will be displayed after the configuration and the submitting of an analysis on an Experiment Package in your home. IDs coming from your datasets will be portrayed as different indicators on pathways depending on the Analysis Mode chosen (see more “Selecting an Analysis Mode”). This page allows you to:</p>
<blockquote>
<div><ul class="simple">
<li><p>Display your datasets on pathways</p></li>
<li><p>Filter pathways by selecting a specific gene/protein/miRNA/TF ID or by multiple IDs</p></li>
<li><p>Select a cellular compartment and display only the IDs localized in it</p></li>
<li><p>Select indicators and display their infos on the clipboard</p></li>
<li><p>Save indicators of interest and download them as a tab-separated text file</p></li>
<li><p>Filter pathways by selecting an “Agreement” between Gene/Proteins and Methylation, miRNAs or TFs</p></li>
</ul>
</div></blockquote>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The “Analysis Results” page is directly accessible after you hit the “Submit” button allowing you to navigate through pathways right away but depending on how much data PathLay needs to process, some indicators may not be displayed correctly from the go and require more time.</p>
</div>
<section id="pathway-navigation">
<span id="usage-results-mapsel"></span><h3>Pathway Navigation<a class="headerlink" href="#pathway-navigation" title="Permalink to this heading"></a></h3>
<p>Pathways available are listed in alphabetical order in the dropdown menu on the top left corner. The pathway name and ID are displayed and the numbers between round parenthesis give you information on how many indicators are loaded on the pathway. Selecting one pathway from this menu will instantly display it with its the indicators.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The pathway selector will display all the pathways available only when no other selections have been performed. Whenever an ID/Agreement/Cellular compartment based selection is performed, this dropdown menu will provide only those pathways that satisfies the parameters of the selection. To return at the initial, default state, simply reset your selections.</p>
</div>
</section>
<section id="indicators">
<span id="usage-results-legend"></span><h3>Indicators<a class="headerlink" href="#indicators" title="Permalink to this heading"></a></h3>
</section>
<section id="highlighting-feature">
<span id="usage-results-highlight"></span><h3>Highlighting Feature<a class="headerlink" href="#highlighting-feature" title="Permalink to this heading"></a></h3>
<p>Two more dropdown menus are located next to the pathway selection dropdown menu and provides an indicator highligthing feature depending on your analysis configuration and the datasets loaded. The combination of these two selectors allows to choose a data type between Genes, Proteins, miRNAs, Metabolites, TFs and Ontologies and display the list of IDs of that data type. Selecting one ID will fill the pathway selector only with the pathway IDs that contains an indicator that represents the ID chosen. For instance, if you are interested in the pathways that displays an indicator for the COX1 gene (which was included in your transcriptomic dataset), simply select “Genes” and then select “COX1”: the pathway selector will now list only those pathways. The behaviour is a little different when selecting a cellular compartment from the Ontologies section, since it will also hide those indicators that do not belong in the cellular compartment. To reset anyone of these selectors simply select the first option provided by them.</p>
</section>
<section id="the-clipboard">
<span id="usage-results-clipboard"></span><h3>The Clipboard<a class="headerlink" href="#the-clipboard" title="Permalink to this heading"></a></h3>
<p>The Clipboard is a freely movable window that will be displayed as soon as the “Analysis Results” page is loaded. It is composed by two boxes, the “Info” box and the “Selected” box. The “Info” box will display the information related to a clicked indicator, this includes every ID that the indicator is portaying, with their differential expression values and other IDs eventually linked. Cliking on the IDs in the “Info” box will open up a page in the online database of reference for that data type:</p>
<blockquote>
<div><ul class="simple">
<li><p>NCBI for genes</p></li>
<li><p>UniProtKB for proteins</p></li>
<li><p>Mirtarbase for miRNAs</p></li>
<li><p>pippo for TFs.</p></li>
</ul>
</div></blockquote>
<p>Once you have clicked an indicator on a pathway and its information is displayed on the Clipboard, you can click the “Add” button below the “Info” box and save the indicator inside the Clipboard’s “Selected” box. It will be listed as a smaller indicator under the name of the pathway from where the selction has been performed. Selecting this smaller indicator in the box will also display its information in the “Info” box and if the “Remove” button is clicked, the indicator will be removed from the “Selected” box. If instead, the “Select” button under the “Selected” box is clicked, the pathway selector will only list pathways that portray the indicator selected. The last feature provided by the Clipboard is the “Download” button which allows to download a tab separated text document listing all the information regarding selected indicators currently populating the “Selected” box.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Closing the Clipboard by clicking the “X” on the top right corner will not delete any selected indicator from the “Selected” box.</p>
</div>
</section>
<section id="the-logical-section">
<span id="usage-results-logical"></span><h3>The Logical Section<a class="headerlink" href="#the-logical-section" title="Permalink to this heading"></a></h3>
<p>This section in the “Analysis Results” page will be displayed when clicking the “Show Logicals” in the top right corner and allows to perform more sophisticate selections either by IDs or by “Agreement”.</p>
<section id="select-by-ids">
<span id="usage-results-logical-byids"></span><h4>Select by IDs<a class="headerlink" href="#select-by-ids" title="Permalink to this heading"></a></h4>
<p>The first row in the Logical Section is composed by a selector that allows the selection of a data type between “Gene”, “Protein”, “miRNA” and “Metabolite”, depending on your analysis configuration. Selecting one data type will pop up another selector with a list of IDs of the data type chosen. Once an ID is selected, it can be added to the selection pool “IDs Selected” positioned on the right of the Logical Section. Every ID added to the pool is represented as a colored box and can be removed from the pool by clicking the “X” button on the right. Colors are assigned considering the data type:</p>
<blockquote>
<div><ul class="simple">
<li><p>Light blue for genes</p></li>
<li><p>Purple for proteins</p></li>
<li><p>Red for miRNAs</p></li>
<li><p>Yellow for Metabolites</p></li>
</ul>
</div></blockquote>
<p>The “Select” button will perform a pathway selection using the pool content as criteria: only those pathways portraying all the IDs in the pool will be displayed in the pathway selector.
The “Reset” button will empty the selection pool and restore the selectors at their default state.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It is possible that loaded pathways cannot satisfy the requirements for the selection made. In this case the pathway selector will become empty. Resetting the selection with the “Reset” button will restore the pathway selector at its default state.</p>
</div>
</section>
<section id="select-by-agreement">
<span id="usage-results-logical-byagreement"></span><h4>Select by Agreement<a class="headerlink" href="#select-by-agreement" title="Permalink to this heading"></a></h4>
<p>The second row in the Logical Section is composed by a series of selectors that allows the setup of the so called “Agreement” which is a feature that lets you select two related data types and select what type of correlation you are interested in. The main data type can be selected with the first selector which provides a choice between “Gene” and “Protein”, and the second data type can be selected with the other selector which allows a choice between “Gene” or “Protein” (depending on your main data type of choice), “miRNA”, “Methylation” or “TF”. After the two data types are selected, the last step to do is to choose a “Positive Agreement” or a “Negative Agreement” with the last selector and hit the “Add” button. The configured agreement will be displayed in the “Agreements Selected” pool positioned under the “IDs Selected” pool. Once the select button is clicked, indicators that don’t satisfy the requirements of the agreement will be hidden, thus if a pathway doesn’t have visible indicators it will not be included in the pathway selector. Multiple agreements can be added to the pool to make the selection more stringent. How various agreements work is reported in the following table:</p>
<table class="colwidths-given docutils align-default" id="id5">
<caption><span class="caption-text">Types of Agreement</span><a class="headerlink" href="#id5" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 20%" />
<col style="width: 40%" />
<col style="width: 40%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"></th>
<th class="head"><p>Gene</p></th>
<th class="head"><p>Protein</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Gene</p></td>
<td></td>
<td><p>Positive Agreement means Positive Correlation</p></td>
</tr>
<tr class="row-odd"><td><p>Protein</p></td>
<td><p>Positive Agreement means Positive Correlation</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>miRNA</p></td>
<td><p>Positive Agreement means Negative Correlation</p></td>
<td><p>Positive Agreement means Negative Correlation</p></td>
</tr>
<tr class="row-odd"><td><p>Methylation</p></td>
<td><p>Positive Agreement means Negative Correlation</p></td>
<td><p>Positive Agreement means Negative Correlation</p></td>
</tr>
<tr class="row-even"><td><p>TF</p></td>
<td><p>Positive Agreement means Positive Correlation</p></td>
<td><p>Positive Agreement means Positive Correlation</p></td>
</tr>
</tbody>
</table>
<p>As an example, let’s suppose we want to examine only the indicators including a gene and a miRNA in a positive agreement: it is only necessary to select “Gene”, “miRNA” and “Positive” from the selectors provided in the Logical section. A positive agreement between a gene and a miRNA is based on a negative correlation, so the indicators highlighted will either have an up-regulated gene and a down-regulated miRNA or a down-regulated gene and an up-regulated miRNA. As a next step, we could also be looking at methylations: easily enough, agreement selections can be stacked so we could also select “Gene”, “Methylation”, “Positive” and highlight only those indicators that satisfy both the agreements.
As before, to remove it the “X” button on the right or the “Reset” button must be clicked.</p>
</section>
</section>
<section id="settings">
<span id="usage-results-settings"></span><h3>Settings<a class="headerlink" href="#settings" title="Permalink to this heading"></a></h3>
<p>The settings section in the “Analysis Results” page will be displayed when clicking the “Show Settings” button and allows to change some graphical parameters for the indicators like size and transparency values.</p>
<section id="transparency-and-size">
<span id="usage-results-settings-transparencyandsize"></span><h4>Transparency and Size<a class="headerlink" href="#transparency-and-size" title="Permalink to this heading"></a></h4>
<p>Transparency and size of the indicators can be increased and decreased with the “Transparency Up”,”Size Up” and “Transparency Down”,”Size Down” buttons respectively.</p>
</section>
<section id="screenshot">
<span id="usage-results-settings-screenshot"></span><h4>Screenshot<a class="headerlink" href="#screenshot" title="Permalink to this heading"></a></h4>
<p>It is possible to take a shot of the currently displayed pathway by clicking the “Open as image (to save..)” button.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Depending on your browser compatibility this feature can behave differently.</p>
</div>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Matteo Ramazzotti Lorenzo Casbarra.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-XXXXXXXXXX"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-XXXXXXXXXX', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>